{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6a8f4c18-2ff9-44fa-9379-25cae4605dee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyter/Somatic_annotation'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9333f94a-a5bd-466f-898e-db894a656e17",
   "metadata": {},
   "source": [
    "# Quick sanity check for correct Parabricks installation and environment \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e1bba9e-ce27-478d-b9b2-a2f174ec0742",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$ nvidia-smi\n",
      "Sun Nov 16 14:42:23 2025       \n",
      "+-----------------------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 550.90.07              Driver Version: 550.90.07      CUDA Version: 12.4     |\n",
      "|-----------------------------------------+------------------------+----------------------+\n",
      "| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |\n",
      "|                                         |                        |               MIG M. |\n",
      "|=========================================+========================+======================|\n",
      "|   0  Tesla T4                       On  |   00000000:00:04.0 Off |                    0 |\n",
      "| N/A   40C    P8              9W /   70W |       1MiB /  15360MiB |      0%      Default |\n",
      "|                                         |                        |                  N/A |\n",
      "+-----------------------------------------+------------------------+----------------------+\n",
      "|   1  Tesla T4                       On  |   00000000:00:05.0 Off |                    0 |\n",
      "| N/A   40C    P8              9W /   70W |       1MiB /  15360MiB |      0%      Default |\n",
      "|                                         |                        |                  N/A |\n",
      "+-----------------------------------------+------------------------+----------------------+\n",
      "|   2  Tesla T4                       On  |   00000000:00:06.0 Off |                    0 |\n",
      "| N/A   42C    P8             10W /   70W |       1MiB /  15360MiB |      0%      Default |\n",
      "|                                         |                        |                  N/A |\n",
      "+-----------------------------------------+------------------------+----------------------+\n",
      "|   3  Tesla T4                       On  |   00000000:00:07.0 Off |                    0 |\n",
      "| N/A   40C    P8             13W /   70W |       1MiB /  15360MiB |      0%      Default |\n",
      "|                                         |                        |                  N/A |\n",
      "+-----------------------------------------+------------------------+----------------------+\n",
      "                                                                                         \n",
      "+-----------------------------------------------------------------------------------------+\n",
      "| Processes:                                                                              |\n",
      "|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |\n",
      "|        ID   ID                                                               Usage      |\n",
      "|=========================================================================================|\n",
      "|  No running processes found                                                             |\n",
      "+-----------------------------------------------------------------------------------------+\n",
      "\n",
      "$ pbrun --version\n",
      "\n",
      "/bin/sh: 1: pbrun: not found\n",
      "\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Command failed with exit code 127",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 12\u001b[0m\n\u001b[1;32m      9\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCommand failed with exit code \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mresult\u001b[38;5;241m.\u001b[39mreturncode\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     11\u001b[0m check_cmd(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnvidia-smi\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 12\u001b[0m \u001b[43mcheck_cmd\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpbrun --version\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[1], line 9\u001b[0m, in \u001b[0;36mcheck_cmd\u001b[0;34m(cmd)\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;28mprint\u001b[39m(result\u001b[38;5;241m.\u001b[39mstderr)\n\u001b[0;32m----> 9\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCommand failed with exit code \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mresult\u001b[38;5;241m.\u001b[39mreturncode\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Command failed with exit code 127"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "def check_cmd(cmd: str):\n",
    "    print(f\"$ {cmd}\")\n",
    "    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)\n",
    "    print(result.stdout)\n",
    "    if result.returncode != 0:\n",
    "        print(result.stderr)\n",
    "        raise RuntimeError(f\"Command failed with exit code {result.returncode}\")\n",
    "\n",
    "check_cmd(\"nvidia-smi\")\n",
    "check_cmd(\"pbrun --version\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a7f6b349-9722-4fab-b603-f382dad0e2b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unable to find image 'nvcr.io/nvidia/clara/parabricks:v4.3.0-1' locally\n",
      "docker: Error response from daemon: unauthorized: <html>\n",
      "<head><title>401 Authorization Required</title></head>\n",
      "<body>\n",
      "<center><h1>401 Authorization Required</h1></center>\n",
      "<hr><center>nginx/1.22.1</center>\n",
      "</body>\n",
      "</html>\n",
      "\n",
      "Run 'docker run --help' for more information\n"
     ]
    }
   ],
   "source": [
    "#Check for local docker image of the version installed \n",
    "!docker run --rm --gpus all nvcr.io/nvidia/clara/parabricks:v4.3.0-1 pbrun --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e119dcae-ce99-4e74-9caf-15cc1053434d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyter/Parabricks_pathogen_Test'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "77d45690-05d3-42f6-b03a-7a55765a074f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.6.0-1: Pulling from nvidia/clara/clara-parabricks\n",
      "\n",
      "\u001b[1B9dea01a6: Already exists \n",
      "\u001b[1B09cfb5af: Already exists \n",
      "\u001b[1B00b97743: Already exists \n",
      "\u001b[1B7ecc44d6: Already exists \n",
      "\u001b[1B15468fcc: Already exists \n",
      "\u001b[1B987b81f9: Pulling fs layer \n",
      "\u001b[1B8c224bf0: Pulling fs layer \n",
      "\u001b[1B247cb961: Pulling fs layer \n",
      "\u001b[1B105830d6: Pulling fs layer \n",
      "\u001b[1Bdb69489a: Pulling fs layer \n",
      "\u001b[1B5f3ce386: Pulling fs layer \n",
      "\u001b[1BDigest: sha256:d0761eb4b9921bc046c53520287316d545eb79feaeb8f22387e9bb5734650447\n",
      "Status: Downloaded newer image for nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1\n",
      "nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1\n"
     ]
    }
   ],
   "source": [
    "!docker pull nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "17cac2b5-0b67-4f83-a446-6eae80c17fe5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "pbrun: 4.6.0-1\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 pbrun --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2f8e6cc9-7540-41ec-b024-b1d660045614",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyter/Parabricks_pathogen_Test'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c810b0ee-a7ff-4689-9f79-e3243fa7b9fa",
   "metadata": {},
   "source": [
    "# Download reference strain E. coli K-12 MG1655 genome and index it "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0cd99b55-32f1-4abc-9f3b-59a7e6558da8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33m  DEPRECATION: Building 'ncbi-datasets-pylib' using the legacy setup.py bdist_wheel mechanism, which will be removed in a future version. pip 25.3 will enforce this behaviour change. A possible replacement is to use the standardized build interface by setting the `--use-pep517` option, (possibly combined with `--no-build-isolation`), or adding a `pyproject.toml` file to the source tree of 'ncbi-datasets-pylib'. Discussion can be found at https://github.com/pypa/pip/issues/6334\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33m  DEPRECATION: Building 'gffutils' using the legacy setup.py bdist_wheel mechanism, which will be removed in a future version. pip 25.3 will enforce this behaviour change. A possible replacement is to use the standardized build interface by setting the `--use-pep517` option, (possibly combined with `--no-build-isolation`), or adding a `pyproject.toml` file to the source tree of 'gffutils'. Discussion can be found at https://github.com/pypa/pip/issues/6334\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33m  WARNING: Failed to remove contents in a temporary directory '/opt/conda/lib/python3.10/site-packages/google/~rotobuf'.\n",
      "  You can safely remove it manually.\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "google-cloud-datastore 1.15.5 requires protobuf<4.0.0dev, but you have protobuf 4.23.0 which is incompatible.\n",
      "kfp 2.5.0 requires protobuf<4,>=3.13.0, but you have protobuf 4.23.0 which is incompatible.\n",
      "kfp-pipeline-spec 0.2.2 requires protobuf<4,>=3.13.0, but you have protobuf 4.23.0 which is incompatible.\u001b[0m\u001b[31m\n",
      "Hit:1 https://download.docker.com/linux/debian bullseye InRelease\n",
      "Hit:2 https://deb.debian.org/debian bullseye InRelease              \n",
      "Hit:3 https://packages.cloud.google.com/apt gcsfuse-bullseye InRelease\n",
      "Get:4 https://nvidia.github.io/libnvidia-container/stable/deb/amd64  InRelease [1477 B]\n",
      "Hit:5 https://deb.debian.org/debian-security bullseye-security InRelease\n",
      "Hit:6 https://deb.debian.org/debian bullseye-updates InRelease\n",
      "Ign:7 https://deb.debian.org/debian bullseye-backports InRelease\n",
      "Hit:8 https://packages.cloud.google.com/apt google-compute-engine-bullseye-stable InRelease\n",
      "Err:9 https://deb.debian.org/debian bullseye-backports Release\n",
      "  404  Not Found [IP: 151.101.2.132 443]\n",
      "Hit:10 https://packages.cloud.google.com/apt cloud-sdk-bullseye InRelease\n",
      "Hit:11 https://packages.cloud.google.com/apt google-fast-socket InRelease\n",
      "Reading package lists... Done\n",
      "E: The repository 'https://deb.debian.org/debian bullseye-backports Release' no longer has a Release file.\n",
      "N: Updating from such a repository can't be done securely, and is therefore disabled by default.\n",
      "N: See apt-secure(8) manpage for repository creation and user configuration details.\n"
     ]
    }
   ],
   "source": [
    "# Install NCBI datasets + unzip (skip if already installed)\n",
    "!pip install --quiet ncbi-datasets-pylib\n",
    "!sudo apt-get update -y && sudo apt-get install -y unzip\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "51628c0b-1fd4-4af8-bcb2-686b5e376ac6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jupyter/Parabricks_pathogen_Test/reference_mg1655\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.10/site-packages/IPython/core/magics/osm.py:417: UserWarning: using dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "cd reference_mg1655"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9fa4e0ee-5884-46e6-b501-15d59cfbbb02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyter/Parabricks_pathogen_Test/reference_mg1655'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4a2ee47-ebb9-4bb6-bab1-b739ef20bcd2",
   "metadata": {},
   "source": [
    "# Running samtools inside PB container to creare .fai index file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2c32ef97-24fd-4650-8931-9f1c52f3d093",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created FASTA index (.fai)\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "REF_DIR = \"/home/jupyter/Parabricks_pathogen_Test/reference_mg1655\"\n",
    "REF_FASTA = \"Ecoli_K12_MG1655_ref.fasta\"\n",
    "IMG = \"nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1\"\n",
    "\n",
    "cmd = f\"\"\"\n",
    "docker run --rm --gpus all \\\n",
    "    -v {REF_DIR}:/data \\\n",
    "    {IMG} \\\n",
    "    samtools faidx /data/{REF_FASTA}\n",
    "\"\"\"\n",
    "\n",
    "subprocess.run(cmd, shell=True, check=True)\n",
    "print(\"Created FASTA index (.fai)\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f05b61c5-c392-4e3d-8e92-42a432dffb08",
   "metadata": {},
   "source": [
    "# GATK installation and run to create .dict file from ref fasta "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b12f48be-22f9-4978-a897-593d7a62e88e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reference FASTA: /home/jupyter/Parabricks_pathogen_Test/reference_mg1655/Ecoli_K12_MG1655_ref.fasta\n",
      "Dictionary will be written to: /home/jupyter/Parabricks_pathogen_Test/reference_mg1655/Ecoli_K12_MG1655_ref.dict\n",
      "\n",
      "GATK not found in PATH. Installing gatk4 via conda...\n",
      "\n",
      "Running: conda install -y -c bioconda gatk4\n",
      "Retrieving notices: done\n",
      "Channels:\n",
      " - bioconda\n",
      " - conda-forge\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "    current version: 25.5.1\n",
      "    latest version: 25.9.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /opt/conda\n",
      "\n",
      "  added / updated specs:\n",
      "    - gatk4\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    alsa-lib-1.2.14            |       hb9d3cd8_0         553 KB  conda-forge\n",
      "    ca-certificates-2025.11.12 |       hbd8a1cb_0         149 KB  conda-forge\n",
      "    cairo-1.18.4               |       h3394656_0         955 KB  conda-forge\n",
      "    certifi-2025.11.12         |     pyhd8ed1ab_0         153 KB  conda-forge\n",
      "    conda-25.7.0               |  py310hff52083_0         949 KB  conda-forge\n",
      "    font-ttf-dejavu-sans-mono-2.37|       hab24e00_0         388 KB  conda-forge\n",
      "    font-ttf-inconsolata-3.000 |       h77eed37_0          94 KB  conda-forge\n",
      "    font-ttf-source-code-pro-2.038|       h77eed37_0         684 KB  conda-forge\n",
      "    font-ttf-ubuntu-0.83       |       h77eed37_3         1.5 MB  conda-forge\n",
      "    fontconfig-2.15.0          |       h7e30c49_1         259 KB  conda-forge\n",
      "    fonts-conda-ecosystem-1    |                0           4 KB  conda-forge\n",
      "    fonts-conda-forge-1        |       hc364b38_1           4 KB  conda-forge\n",
      "    freetype-2.14.1            |       ha770c72_0         169 KB  conda-forge\n",
      "    gatk4-4.6.2.0              |  py310hdfd78af_1       374.0 MB  bioconda\n",
      "    giflib-5.2.2               |       hd590300_0          75 KB  conda-forge\n",
      "    graphite2-1.3.14           |       hecca717_2          97 KB  conda-forge\n",
      "    harfbuzz-11.2.1            |       h3beb420_0         1.7 MB  conda-forge\n",
      "    lcms2-2.17                 |       h717163a_0         242 KB  conda-forge\n",
      "    lerc-4.0.0                 |       h0aef613_1         258 KB  conda-forge\n",
      "    libcups-2.3.3              |       hb8b1518_5         4.3 MB  conda-forge\n",
      "    libdeflate-1.25            |       h17f619e_0          72 KB  conda-forge\n",
      "    libfreetype-2.14.1         |       ha770c72_0           7 KB  conda-forge\n",
      "    libfreetype6-2.14.1        |       h73754d4_0         378 KB  conda-forge\n",
      "    libglib-2.86.0             |       h1fed272_0         3.8 MB  conda-forge\n",
      "    libjpeg-turbo-3.1.2        |       hb03c661_0         619 KB  conda-forge\n",
      "    libpng-1.6.50              |       h421ea60_1         310 KB  conda-forge\n",
      "    libtiff-4.7.1              |       h9d88235_1         425 KB  conda-forge\n",
      "    libwebp-base-1.6.0         |       hd42ef1d_0         419 KB  conda-forge\n",
      "    libxcb-1.17.0              |       h8a09558_0         387 KB  conda-forge\n",
      "    openjdk-23.0.2             |       h53dfc1b_2       181.4 MB  conda-forge\n",
      "    openssl-3.6.0              |       h26f9b46_0         3.0 MB  conda-forge\n",
      "    pcre2-10.46                |       h1321c63_0         1.2 MB  conda-forge\n",
      "    pixman-0.46.4              |       h54a6638_1         440 KB  conda-forge\n",
      "    pthread-stubs-0.4          |    hb9d3cd8_1002           8 KB  conda-forge\n",
      "    requests-2.32.5            |     pyhd8ed1ab_0          58 KB  conda-forge\n",
      "    urllib3-2.5.0              |     pyhd8ed1ab_0          99 KB  conda-forge\n",
      "    xorg-libice-1.1.2          |       hb9d3cd8_0          57 KB  conda-forge\n",
      "    xorg-libsm-1.2.6           |       he73a12e_0          27 KB  conda-forge\n",
      "    xorg-libx11-1.8.12         |       h4f16b4b_0         816 KB  conda-forge\n",
      "    xorg-libxau-1.0.12         |       hb03c661_1          15 KB  conda-forge\n",
      "    xorg-libxdmcp-1.1.5        |       hb03c661_1          20 KB  conda-forge\n",
      "    xorg-libxext-1.3.6         |       hb9d3cd8_0          49 KB  conda-forge\n",
      "    xorg-libxfixes-6.0.2       |       hb03c661_0          20 KB  conda-forge\n",
      "    xorg-libxi-1.8.2           |       hb9d3cd8_0          46 KB  conda-forge\n",
      "    xorg-libxrandr-1.5.4       |       hb9d3cd8_0          29 KB  conda-forge\n",
      "    xorg-libxrender-0.9.12     |       hb9d3cd8_0          32 KB  conda-forge\n",
      "    xorg-libxt-1.3.1           |       hb9d3cd8_0         371 KB  conda-forge\n",
      "    xorg-libxtst-1.2.5         |       hb9d3cd8_3          32 KB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:       580.4 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  alsa-lib           conda-forge/linux-64::alsa-lib-1.2.14-hb9d3cd8_0 \n",
      "  cairo              conda-forge/linux-64::cairo-1.18.4-h3394656_0 \n",
      "  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0 \n",
      "  font-ttf-inconsol~ conda-forge/noarch::font-ttf-inconsolata-3.000-h77eed37_0 \n",
      "  font-ttf-source-c~ conda-forge/noarch::font-ttf-source-code-pro-2.038-h77eed37_0 \n",
      "  font-ttf-ubuntu    conda-forge/noarch::font-ttf-ubuntu-0.83-h77eed37_3 \n",
      "  fontconfig         conda-forge/linux-64::fontconfig-2.15.0-h7e30c49_1 \n",
      "  fonts-conda-ecosy~ conda-forge/noarch::fonts-conda-ecosystem-1-0 \n",
      "  fonts-conda-forge  conda-forge/noarch::fonts-conda-forge-1-hc364b38_1 \n",
      "  freetype           conda-forge/linux-64::freetype-2.14.1-ha770c72_0 \n",
      "  gatk4              bioconda/noarch::gatk4-4.6.2.0-py310hdfd78af_1 \n",
      "  giflib             conda-forge/linux-64::giflib-5.2.2-hd590300_0 \n",
      "  graphite2          conda-forge/linux-64::graphite2-1.3.14-hecca717_2 \n",
      "  harfbuzz           conda-forge/linux-64::harfbuzz-11.2.1-h3beb420_0 \n",
      "  lcms2              conda-forge/linux-64::lcms2-2.17-h717163a_0 \n",
      "  lerc               conda-forge/linux-64::lerc-4.0.0-h0aef613_1 \n",
      "  libcups            conda-forge/linux-64::libcups-2.3.3-hb8b1518_5 \n",
      "  libdeflate         conda-forge/linux-64::libdeflate-1.25-h17f619e_0 \n",
      "  libfreetype        conda-forge/linux-64::libfreetype-2.14.1-ha770c72_0 \n",
      "  libfreetype6       conda-forge/linux-64::libfreetype6-2.14.1-h73754d4_0 \n",
      "  libglib            conda-forge/linux-64::libglib-2.86.0-h1fed272_0 \n",
      "  libjpeg-turbo      conda-forge/linux-64::libjpeg-turbo-3.1.2-hb03c661_0 \n",
      "  libpng             conda-forge/linux-64::libpng-1.6.50-h421ea60_1 \n",
      "  libtiff            conda-forge/linux-64::libtiff-4.7.1-h9d88235_1 \n",
      "  libwebp-base       conda-forge/linux-64::libwebp-base-1.6.0-hd42ef1d_0 \n",
      "  libxcb             conda-forge/linux-64::libxcb-1.17.0-h8a09558_0 \n",
      "  openjdk            conda-forge/linux-64::openjdk-23.0.2-h53dfc1b_2 \n",
      "  pcre2              conda-forge/linux-64::pcre2-10.46-h1321c63_0 \n",
      "  pixman             conda-forge/linux-64::pixman-0.46.4-h54a6638_1 \n",
      "  pthread-stubs      conda-forge/linux-64::pthread-stubs-0.4-hb9d3cd8_1002 \n",
      "  requests           conda-forge/noarch::requests-2.32.5-pyhd8ed1ab_0 \n",
      "  urllib3            conda-forge/noarch::urllib3-2.5.0-pyhd8ed1ab_0 \n",
      "  xorg-libice        conda-forge/linux-64::xorg-libice-1.1.2-hb9d3cd8_0 \n",
      "  xorg-libsm         conda-forge/linux-64::xorg-libsm-1.2.6-he73a12e_0 \n",
      "  xorg-libx11        conda-forge/linux-64::xorg-libx11-1.8.12-h4f16b4b_0 \n",
      "  xorg-libxau        conda-forge/linux-64::xorg-libxau-1.0.12-hb03c661_1 \n",
      "  xorg-libxdmcp      conda-forge/linux-64::xorg-libxdmcp-1.1.5-hb03c661_1 \n",
      "  xorg-libxext       conda-forge/linux-64::xorg-libxext-1.3.6-hb9d3cd8_0 \n",
      "  xorg-libxfixes     conda-forge/linux-64::xorg-libxfixes-6.0.2-hb03c661_0 \n",
      "  xorg-libxi         conda-forge/linux-64::xorg-libxi-1.8.2-hb9d3cd8_0 \n",
      "  xorg-libxrandr     conda-forge/linux-64::xorg-libxrandr-1.5.4-hb9d3cd8_0 \n",
      "  xorg-libxrender    conda-forge/linux-64::xorg-libxrender-0.9.12-hb9d3cd8_0 \n",
      "  xorg-libxt         conda-forge/linux-64::xorg-libxt-1.3.1-hb9d3cd8_0 \n",
      "  xorg-libxtst       conda-forge/linux-64::xorg-libxtst-1.2.5-hb9d3cd8_3 \n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "  ca-certificates                      2025.4.26-hbd8a1cb_0 --> 2025.11.12-hbd8a1cb_0 \n",
      "  certifi                            2025.4.26-pyhd8ed1ab_0 --> 2025.11.12-pyhd8ed1ab_0 \n",
      "  conda                              25.5.1-py310hff52083_0 --> 25.7.0-py310hff52083_0 \n",
      "  openssl                                  3.5.0-h7b32b05_1 --> 3.6.0-h26f9b46_0 \n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  |            |   0% \n",
      "openjdk-23.0.2       | 181.4 MB  |            |   0% \u001b[A\n",
      "\n",
      "libcups-2.3.3        | 4.3 MB    |            |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "libglib-2.86.0       | 3.8 MB    |            |   0% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "openssl-3.6.0        | 3.0 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "harfbuzz-11.2.1      | 1.7 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-ubuntu-0.83 | 1.5 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pcre2-10.46          | 1.2 MB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "cairo-1.18.4         | 955 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "conda-25.7.0         | 949 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xorg-libx11-1.8.12   | 816 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-source-code | 684 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libjpeg-turbo-3.1.2  | 619 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "alsa-lib-1.2.14      | 553 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pixman-0.46.4        | 440 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libtiff-4.7.1        | 425 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libwebp-base-1.6.0   | 419 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-dejavu-sans | 388 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libxcb-1.17.0        | 387 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "openssl-3.6.0        | 3.0 MB    | 4          |   4% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  |            |   0% \u001b[A\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  |            |   0% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "libcups-2.3.3        | 4.3 MB    |            |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "openssl-3.6.0        | 3.0 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | 2          |   2% \u001b[A\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  |            |   1% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "libcups-2.3.3        | 4.3 MB    | #######5   |  75% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "harfbuzz-11.2.1      | 1.7 MB    |            |   1% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | 2          |   2% \u001b[A\n",
      "\n",
      "\n",
      "libglib-2.86.0       | 3.8 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "harfbuzz-11.2.1      | 1.7 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "harfbuzz-11.2.1      | 1.7 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "libcups-2.3.3        | 4.3 MB    | ########## | 100% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | 2          |   3% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | 6          |   7% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pcre2-10.46          | 1.2 MB    | 1          |   1% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "cairo-1.18.4         | 955 KB    | 1          |   2% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-ubuntu-0.83 | 1.5 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-ubuntu-0.83 | 1.5 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | 3          |   4% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | 8          |   9% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pcre2-10.46          | 1.2 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "conda-25.7.0         | 949 KB    | 1          |   2% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-source-code | 684 KB    | 2          |   2% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | 5          |   5% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | #1         |  11% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "conda-25.7.0         | 949 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xorg-libx11-1.8.12   | 816 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-source-code | 684 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libjpeg-turbo-3.1.2  | 619 KB    | 2          |   3% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "alsa-lib-1.2.14      | 553 KB    | 2          |   3% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | 6          |   6% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libjpeg-turbo-3.1.2  | 619 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pixman-0.46.4        | 440 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "alsa-lib-1.2.14      | 553 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | #3         |  13% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "harfbuzz-11.2.1      | 1.7 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libwebp-base-1.6.0   | 419 KB    | 3          |   4% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libtiff-4.7.1        | 425 KB    | 3          |   4% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-dejavu-sans | 388 KB    | 4          |   4% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libwebp-base-1.6.0   | 419 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "openssl-3.6.0        | 3.0 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "openssl-3.6.0        | 3.0 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | 6          |   7% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libtiff-4.7.1        | 425 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-dejavu-sans | 388 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libxcb-1.17.0        | 387 KB    | 4          |   4% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libxcb-1.17.0        | 387 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | 7          |   8% [A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | #6         |  17% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | 8          |   9% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | 9          |   9% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #          |  10% \u001b[A\n",
      "\n",
      "libcups-2.3.3        | 4.3 MB    | ########## | 100% \u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | ##2        |  22% \u001b[A\n",
      "\n",
      "\n",
      "libglib-2.86.0       | 3.8 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "cairo-1.18.4         | 955 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "cairo-1.18.4         | 955 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #1         |  11% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #1         |  12% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #3         |  13% \u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | ##9        |  29% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #4         |  14% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #4         |  15% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #5         |  15% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xorg-libx11-1.8.12   | 816 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xorg-libx11-1.8.12   | 816 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pcre2-10.46          | 1.2 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pcre2-10.46          | 1.2 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #6         |  16% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #7         |  17% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libjpeg-turbo-3.1.2  | 619 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libjpeg-turbo-3.1.2  | 619 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | ###8       |  39% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-source-code | 684 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #8         |  18% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pixman-0.46.4        | 440 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pixman-0.46.4        | 440 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "conda-25.7.0         | 949 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "conda-25.7.0         | 949 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #8         |  19% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libwebp-base-1.6.0   | 419 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libwebp-base-1.6.0   | 419 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | ####2      |  42% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "font-ttf-dejavu-sans | 388 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #9         |  19% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | ####4      |  44% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libtiff-4.7.1        | 425 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##         |  20% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | ####5      |  46% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "alsa-lib-1.2.14      | 553 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##         |  21% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "openjdk-23.0.2       | 181.4 MB  | ####7      |  47% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libxcb-1.17.0        | 387 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##1        |  22% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##2        |  22% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##2        |  23% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##3        |  24% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##4        |  24% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##5        |  25% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##5        |  26% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##7        |  27% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ##9        |  30% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ###1       |  32% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ###3       |  34% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ###6       |  36% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ####       |  41% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ####2      |  43% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ####5      |  45% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ####7      |  48% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #####      |  51% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #####3     |  53% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #####5     |  56% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #####8     |  58% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | #######1   |  71% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ########6  |  87% \u001b[A\n",
      "gatk4-4.6.2.0        | 374.0 MB  | ########## | 100% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     A\u001b[A\n",
      "                                                     \u001b[A\n",
      "\n",
      "                                                     \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\u001b[A\n",
      "\n",
      "\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\u001b[A\n",
      "\n",
      "\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A done\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "\n",
      "GATK successfully installed at: /opt/conda/bin/gatk\n",
      "\n",
      "Running GATK CreateSequenceDictionary...\n",
      "Command:\n",
      "gatk CreateSequenceDictionary -R /home/jupyter/Parabricks_pathogen_Test/reference_mg1655/Ecoli_K12_MG1655_ref.fasta -O /home/jupyter/Parabricks_pathogen_Test/reference_mg1655/Ecoli_K12_MG1655_ref.dict\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using GATK jar /opt/conda/share/gatk4-4.6.2.0-1/gatk-package-4.6.2.0-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /opt/conda/share/gatk4-4.6.2.0-1/gatk-package-4.6.2.0-local.jar CreateSequenceDictionary -R /home/jupyter/Parabricks_pathogen_Test/reference_mg1655/Ecoli_K12_MG1655_ref.fasta -O /home/jupyter/Parabricks_pathogen_Test/reference_mg1655/Ecoli_K12_MG1655_ref.dict\n",
      "18:12:37.348 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/opt/conda/share/gatk4-4.6.2.0-1/gatk-package-4.6.2.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Sun Nov 16 18:12:37 UTC 2025] CreateSequenceDictionary --OUTPUT /home/jupyter/Parabricks_pathogen_Test/reference_mg1655/Ecoli_K12_MG1655_ref.dict --REFERENCE /home/jupyter/Parabricks_pathogen_Test/reference_mg1655/Ecoli_K12_MG1655_ref.fasta --TRUNCATE_NAMES_AT_WHITESPACE true --NUM_SEQUENCES 2147483647 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "[Sun Nov 16 18:12:37 UTC 2025] Executing as jupyter@ac-genomics on Linux 5.10.0-36-cloud-amd64 amd64; OpenJDK 64-Bit Server VM 23.0.2-internal-adhoc.conda.src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: Version:4.6.2.0\n",
      "[Sun Nov 16 18:12:37 UTC 2025] picard.sam.CreateSequenceDictionary done. Elapsed time: 0.00 minutes.\n",
      "Runtime.totalMemory()=167772160\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tool returned:\n",
      "0\n",
      " Sequence dictionary created successfully:\n",
      "/home/jupyter/Parabricks_pathogen_Test/reference_mg1655/Ecoli_K12_MG1655_ref.dict\n",
      "\n",
      "Directory contents:\n",
      "   Ecoli_K12_MG1655_ref.dict\n",
      "   Ecoli_K12_MG1655_ref.fasta.fai\n",
      "   Ecoli_K12_MG1655_ref.fasta\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "\n",
    "# ---- EDIT THESE ONLY IF YOUR PATHS ARE DIFFERENT ----\n",
    "REF_DIR = \"/home/jupyter/Parabricks_pathogen_Test/reference_mg1655\"\n",
    "REF_FASTA = \"Ecoli_K12_MG1655_ref.fasta\"\n",
    "# -----------------------------------------------------\n",
    "\n",
    "ref_path = os.path.join(REF_DIR, REF_FASTA)\n",
    "dict_path = os.path.join(REF_DIR, \"Ecoli_K12_MG1655_ref.dict\")\n",
    "\n",
    "if not os.path.exists(ref_path):\n",
    "    raise FileNotFoundError(f\"Reference FASTA not found: {ref_path}\")\n",
    "\n",
    "print(f\"Reference FASTA: {ref_path}\")\n",
    "print(f\"Dictionary will be written to: {dict_path}\\n\")\n",
    "\n",
    "# 1) Check if gatk is already available\n",
    "gatk_path = shutil.which(\"gatk\")\n",
    "\n",
    "if gatk_path:\n",
    "    print(f\"Found existing GATK at: {gatk_path}\")\n",
    "else:\n",
    "    print(\"GATK not found in PATH. Installing gatk4 via conda...\\n\")\n",
    "    # Try installing gatk4 with conda (requires conda in this environment)\n",
    "    conda_path = shutil.which(\"conda\")\n",
    "    if not conda_path:\n",
    "        raise RuntimeError(\n",
    "            \"conda is not available, cannot auto-install GATK. \"\n",
    "            \"Please install GATK manually or add it to PATH.\"\n",
    "        )\n",
    "    # Install gatk4 into the current conda env\n",
    "    install_cmd = \"conda install -y -c bioconda gatk4\"\n",
    "    print(f\"Running: {install_cmd}\")\n",
    "    subprocess.run(install_cmd, shell=True, check=True)\n",
    "    \n",
    "    # Re-check gatk in PATH\n",
    "    gatk_path = shutil.which(\"gatk\")\n",
    "    if not gatk_path:\n",
    "        raise RuntimeError(\n",
    "            \"GATK installation via conda completed but 'gatk' is still not in PATH. \"\n",
    "            \"You may need to restart the kernel or ensure the conda env is activated.\"\n",
    "        )\n",
    "    print(f\"\\nGATK successfully installed at: {gatk_path}\")\n",
    "\n",
    "# 2) Run GATK CreateSequenceDictionary\n",
    "print(\"\\nRunning GATK CreateSequenceDictionary...\")\n",
    "cmd = (\n",
    "    f\"gatk CreateSequenceDictionary \"\n",
    "    f\"-R {ref_path} \"\n",
    "    f\"-O {dict_path}\"\n",
    ")\n",
    "print(f\"Command:\\n{cmd}\\n\")\n",
    "\n",
    "subprocess.run(cmd, shell=True, check=True)\n",
    "\n",
    "# 3) Confirm output\n",
    "if os.path.exists(dict_path):\n",
    "    print(\" Sequence dictionary created successfully:\")\n",
    "    print(dict_path)\n",
    "    print(\"\\nDirectory contents:\")\n",
    "    for f in os.listdir(REF_DIR):\n",
    "        print(\"  \", f)\n",
    "else:\n",
    "    raise RuntimeError(\"GATK ran without error but the .dict file was not found.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "02e8a531-d178-45eb-993d-fa6fbd660533",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyter/Parabricks_pathogen_Test/reference_mg1655'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5987d6de-f5df-417e-b33e-b72734323a01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entrez Direct not found. Installing via conda (bioconda/entrez-direct)...\n",
      "\n",
      "$ conda install -y -c bioconda entrez-direct\n",
      "Channels:\n",
      " - bioconda\n",
      " - conda-forge\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /opt/conda\n",
      "\n",
      "  added / updated specs:\n",
      "    - entrez-direct\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    entrez-direct-24.0         |       he881be0_0        16.3 MB  bioconda\n",
      "    libidn2-2.3.8              |       hfac485b_1         136 KB  conda-forge\n",
      "    libunistring-0.9.10        |       h7f98852_0         1.4 MB  conda-forge\n",
      "    wget-1.21.4                |       hda4d442_0         752 KB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        18.6 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  entrez-direct      bioconda/linux-64::entrez-direct-24.0-he881be0_0 \n",
      "  libidn2            conda-forge/linux-64::libidn2-2.3.8-hfac485b_1 \n",
      "  libunistring       conda-forge/linux-64::libunistring-0.9.10-h7f98852_0 \n",
      "  wget               conda-forge/linux-64::wget-1.21.4-hda4d442_0 \n",
      "\n",
      "\n",
      "\n",
      "entrez-direct-24.0   | 16.3 MB   |            |   0% \n",
      "libunistring-0.9.10  | 1.4 MB    |            |   0% \u001b[A\n",
      "\n",
      "wget-1.21.4          | 752 KB    |            |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "libidn2-2.3.8        | 136 KB    |            |   0% \u001b[A\u001b[A\u001b[A"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "    current version: 25.7.0\n",
      "    latest version: 25.9.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "libunistring-0.9.10  | 1.4 MB    | 1          |   1% \u001b[A\n",
      "\n",
      "\n",
      "libidn2-2.3.8        | 136 KB    | #1         |  12% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "wget-1.21.4          | 752 KB    | 2          |   2% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "libidn2-2.3.8        | 136 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "libidn2-2.3.8        | 136 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "entrez-direct-24.0   | 16.3 MB   |            |   0% \u001b[A\u001b[A\n",
      "libunistring-0.9.10  | 1.4 MB    | ########## | 100% \u001b[A\n",
      "\n",
      "wget-1.21.4          | 752 KB    | ########## | 100% \u001b[A\u001b[A\n",
      "\n",
      "entrez-direct-24.0   | 16.3 MB   | #####7     |  58% \u001b[A\u001b[A\n",
      "libunistring-0.9.10  | 1.4 MB    | ########## | 100% \u001b[A\n",
      "                                                     \u001b[A\n",
      "                                                     \u001b[A\n",
      "\n",
      "                                                     \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A done\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "SRA Toolkit not found. Installing via conda (bioconda/sra-tools)...\n",
      "\n",
      "$ conda install -y -c bioconda sra-tools\n",
      "Channels:\n",
      " - bioconda\n",
      " - conda-forge\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "    current version: 25.7.0\n",
      "    latest version: 25.9.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /opt/conda\n",
      "\n",
      "  added / updated specs:\n",
      "    - sra-tools\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    curl-8.13.0                |       h332b0f4_0         178 KB  conda-forge\n",
      "    ncbi-vdb-3.2.1             |       h9948957_0        10.6 MB  bioconda\n",
      "    ossuuid-1.6.2              |    h5888daf_1001          53 KB  conda-forge\n",
      "    perl-5.32.1                | 7_hd590300_perl5        12.7 MB  conda-forge\n",
      "    perl-alien-build-2.84      | pl5321h7b50bb2_1         195 KB  bioconda\n",
      "    perl-alien-libxml2-0.17    | pl5321h577a1d6_1          14 KB  bioconda\n",
      "    perl-business-isbn-3.007   | pl5321hdfd78af_0          18 KB  bioconda\n",
      "    perl-business-isbn-data-20210112.006| pl5321hdfd78af_0          21 KB  bioconda\n",
      "    perl-capture-tiny-0.48     | pl5321hdfd78af_2          19 KB  bioconda\n",
      "    perl-carp-1.38             | pl5321hdfd78af_4          17 KB  bioconda\n",
      "    perl-constant-1.33         | pl5321hdfd78af_2          12 KB  bioconda\n",
      "    perl-data-dumper-2.183     | pl5321hec16e2b_1          38 KB  bioconda\n",
      "    perl-encode-3.19           | pl5321hec16e2b_1         2.1 MB  bioconda\n",
      "    perl-exporter-5.72         | pl5321hdfd78af_2          16 KB  bioconda\n",
      "    perl-extutils-makemaker-7.70| pl5321hd8ed1ab_0         154 KB  conda-forge\n",
      "    perl-ffi-checklib-0.28     | pl5321hdfd78af_0          16 KB  bioconda\n",
      "    perl-file-chdir-0.1010     | pl5321hdfd78af_3          13 KB  bioconda\n",
      "    perl-file-path-2.18        | pl5321hd8ed1ab_0          22 KB  conda-forge\n",
      "    perl-file-temp-0.2304      | pl5321hd8ed1ab_0          31 KB  conda-forge\n",
      "    perl-file-which-1.24       | pl5321hd8ed1ab_0          17 KB  conda-forge\n",
      "    perl-importer-0.026        | pl5321hdfd78af_0          23 KB  bioconda\n",
      "    perl-mime-base64-3.16      | pl5321hec16e2b_2          17 KB  bioconda\n",
      "    perl-parent-0.236          | pl5321hdfd78af_2           8 KB  bioconda\n",
      "    perl-path-tiny-0.122       | pl5321hdfd78af_0          40 KB  bioconda\n",
      "    perl-pathtools-3.75        | pl5321hec16e2b_3          42 KB  bioconda\n",
      "    perl-scope-guard-0.21      | pl5321hdfd78af_3          10 KB  bioconda\n",
      "    perl-sub-info-0.002        | pl5321hdfd78af_1          10 KB  bioconda\n",
      "    perl-term-table-0.027      | pl5321hdfd78af_0          24 KB  bioconda\n",
      "    perl-test2-suite-0.000163  | pl5321hdfd78af_0         208 KB  bioconda\n",
      "    perl-uri-5.12              | pl5321hdfd78af_0          59 KB  bioconda\n",
      "    perl-xml-libxml-2.0210     | pl5321hf886d80_0         271 KB  bioconda\n",
      "    perl-xml-namespacesupport-1.12| pl5321hdfd78af_1          14 KB  bioconda\n",
      "    perl-xml-sax-1.02          | pl5321hdfd78af_1          37 KB  bioconda\n",
      "    perl-xml-sax-base-1.09     | pl5321hdfd78af_1          21 KB  bioconda\n",
      "    sra-tools-3.2.1            |       h4304569_1        61.0 MB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        88.0 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  curl               conda-forge/linux-64::curl-8.13.0-h332b0f4_0 \n",
      "  ncbi-vdb           bioconda/linux-64::ncbi-vdb-3.2.1-h9948957_0 \n",
      "  ossuuid            conda-forge/linux-64::ossuuid-1.6.2-h5888daf_1001 \n",
      "  perl               conda-forge/linux-64::perl-5.32.1-7_hd590300_perl5 \n",
      "  perl-alien-build   bioconda/linux-64::perl-alien-build-2.84-pl5321h7b50bb2_1 \n",
      "  perl-alien-libxml2 bioconda/linux-64::perl-alien-libxml2-0.17-pl5321h577a1d6_1 \n",
      "  perl-business-isbn bioconda/noarch::perl-business-isbn-3.007-pl5321hdfd78af_0 \n",
      "  perl-business-isb~ bioconda/noarch::perl-business-isbn-data-20210112.006-pl5321hdfd78af_0 \n",
      "  perl-capture-tiny  bioconda/noarch::perl-capture-tiny-0.48-pl5321hdfd78af_2 \n",
      "  perl-carp          bioconda/noarch::perl-carp-1.38-pl5321hdfd78af_4 \n",
      "  perl-constant      bioconda/noarch::perl-constant-1.33-pl5321hdfd78af_2 \n",
      "  perl-data-dumper   bioconda/linux-64::perl-data-dumper-2.183-pl5321hec16e2b_1 \n",
      "  perl-encode        bioconda/linux-64::perl-encode-3.19-pl5321hec16e2b_1 \n",
      "  perl-exporter      bioconda/noarch::perl-exporter-5.72-pl5321hdfd78af_2 \n",
      "  perl-extutils-mak~ conda-forge/noarch::perl-extutils-makemaker-7.70-pl5321hd8ed1ab_0 \n",
      "  perl-ffi-checklib  bioconda/noarch::perl-ffi-checklib-0.28-pl5321hdfd78af_0 \n",
      "  perl-file-chdir    bioconda/noarch::perl-file-chdir-0.1010-pl5321hdfd78af_3 \n",
      "  perl-file-path     conda-forge/noarch::perl-file-path-2.18-pl5321hd8ed1ab_0 \n",
      "  perl-file-temp     conda-forge/noarch::perl-file-temp-0.2304-pl5321hd8ed1ab_0 \n",
      "  perl-file-which    conda-forge/noarch::perl-file-which-1.24-pl5321hd8ed1ab_0 \n",
      "  perl-importer      bioconda/noarch::perl-importer-0.026-pl5321hdfd78af_0 \n",
      "  perl-mime-base64   bioconda/linux-64::perl-mime-base64-3.16-pl5321hec16e2b_2 \n",
      "  perl-parent        bioconda/noarch::perl-parent-0.236-pl5321hdfd78af_2 \n",
      "  perl-path-tiny     bioconda/noarch::perl-path-tiny-0.122-pl5321hdfd78af_0 \n",
      "  perl-pathtools     bioconda/linux-64::perl-pathtools-3.75-pl5321hec16e2b_3 \n",
      "  perl-scope-guard   bioconda/noarch::perl-scope-guard-0.21-pl5321hdfd78af_3 \n",
      "  perl-sub-info      bioconda/noarch::perl-sub-info-0.002-pl5321hdfd78af_1 \n",
      "  perl-term-table    bioconda/noarch::perl-term-table-0.027-pl5321hdfd78af_0 \n",
      "  perl-test2-suite   bioconda/noarch::perl-test2-suite-0.000163-pl5321hdfd78af_0 \n",
      "  perl-uri           bioconda/noarch::perl-uri-5.12-pl5321hdfd78af_0 \n",
      "  perl-xml-libxml    bioconda/linux-64::perl-xml-libxml-2.0210-pl5321hf886d80_0 \n",
      "  perl-xml-namespac~ bioconda/noarch::perl-xml-namespacesupport-1.12-pl5321hdfd78af_1 \n",
      "  perl-xml-sax       bioconda/noarch::perl-xml-sax-1.02-pl5321hdfd78af_1 \n",
      "  perl-xml-sax-base  bioconda/noarch::perl-xml-sax-base-1.09-pl5321hdfd78af_1 \n",
      "  sra-tools          bioconda/linux-64::sra-tools-3.2.1-h4304569_1 \n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   |            |   0% \n",
      "perl-5.32.1          | 12.7 MB   |            |   0% \u001b[A\n",
      "\n",
      "ncbi-vdb-3.2.1       | 10.6 MB   |            |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "perl-encode-3.19     | 2.1 MB    |            |   0% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "perl-xml-libxml-2.02 | 271 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-test2-suite-0.0 | 208 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-alien-build-2.8 | 195 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "curl-8.13.0          | 178 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-extutils-makema | 154 KB    |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-uri-5.12        | 59 KB     |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ossuuid-1.6.2        | 53 KB     |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-pathtools-3.75  | 42 KB     |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-path-tiny-0.122 | 40 KB     |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-data-dumper-2.1 | 38 KB     |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-xml-sax-1.02    | 37 KB     |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-file-temp-0.230 | 31 KB     |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-term-table-0.02 | 24 KB     |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-importer-0.026  | 23 KB     |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-file-path-2.18  | 22 KB     |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "perl-encode-3.19     | 2.1 MB    |            |   1% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "ncbi-vdb-3.2.1       | 10.6 MB   |            |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   |            |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "perl-xml-libxml-2.02 | 271 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "perl-encode-3.19     | 2.1 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\n",
      "perl-5.32.1          | 12.7 MB   |            |   0% \u001b[A\n",
      "\n",
      "ncbi-vdb-3.2.1       | 10.6 MB   | ####1      |  41% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | 5          |   6% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-test2-suite-0.0 | 208 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-alien-build-2.8 | 195 KB    | 8          |   8% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "perl-5.32.1          | 12.7 MB   | #5         |  15% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-alien-build-2.8 | 195 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "curl-8.13.0          | 178 KB    | 8          |   9% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | #          |  10% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "curl-8.13.0          | 178 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "perl-5.32.1          | 12.7 MB   | ####1      |  41% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-extutils-makema | 154 KB    | #          |  10% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-extutils-makema | 154 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-uri-5.12        | 59 KB     | ##6        |  27% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "ncbi-vdb-3.2.1       | 10.6 MB   | #########6 |  97% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-uri-5.12        | 59 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "perl-xml-libxml-2.02 | 271 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | #5         |  15% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ossuuid-1.6.2        | 53 KB     | ###        |  30% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "perl-5.32.1          | 12.7 MB   | ######7    |  68% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ossuuid-1.6.2        | 53 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-pathtools-3.75  | 42 KB     | ###7       |  38% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | ##1        |  21% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-path-tiny-0.122 | 40 KB     | ####       |  40% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-path-tiny-0.122 | 40 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "perl-5.32.1          | 12.7 MB   | #########  |  90% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-data-dumper-2.1 | 38 KB     | ####2      |  43% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | ##6        |  27% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-xml-sax-1.02    | 37 KB     | ####2      |  43% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-xml-sax-1.02    | 37 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-file-temp-0.230 | 31 KB     | #####1     |  52% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-file-temp-0.230 | 31 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-term-table-0.02 | 24 KB     | ######6    |  66% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | ###2       |  33% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-importer-0.026  | 23 KB     | ######9    |  70% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-importer-0.026  | 23 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "ncbi-vdb-3.2.1       | 10.6 MB   | ########## | 100% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-file-path-2.18  | 22 KB     | #######3   |  73% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-file-path-2.18  | 22 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | ####       |  40% [A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-test2-suite-0.0 | 208 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-test2-suite-0.0 | 208 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "curl-8.13.0          | 178 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "curl-8.13.0          | 178 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-alien-build-2.8 | 195 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | ####8      |  48% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "perl-5.32.1          | 12.7 MB   | ########## | 100% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-extutils-makema | 154 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-extutils-makema | 154 KB    | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ossuuid-1.6.2        | 53 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "ossuuid-1.6.2        | 53 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-uri-5.12        | 59 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | #####6     |  56% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-pathtools-3.75  | 42 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-pathtools-3.75  | 42 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-path-tiny-0.122 | 40 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-path-tiny-0.122 | 40 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-data-dumper-2.1 | 38 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-data-dumper-2.1 | 38 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-file-temp-0.230 | 31 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-file-temp-0.230 | 31 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "perl-encode-3.19     | 2.1 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "perl-encode-3.19     | 2.1 MB    | ########## | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-xml-sax-1.02    | 37 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-xml-sax-1.02    | 37 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-importer-0.026  | 23 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | ######5    |  65% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-term-table-0.02 | 24 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-term-table-0.02 | 24 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-file-path-2.18  | 22 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "perl-file-path-2.18  | 22 KB     | ########## | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | ########## | 100% [A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "perl-5.32.1          | 12.7 MB   | ########## | 100% \u001b[A\n",
      "\n",
      "sra-tools-3.2.1      | 61.0 MB   | ########## | 100% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     A\u001b[A\n",
      "                                                     \u001b[A\n",
      "\n",
      "                                                     \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                     \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\u001b[A\n",
      "\n",
      "\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A done\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "\n",
      "Fetching SRA RunInfo for BioProject PRJNA528413 ...\n",
      "\n",
      "$ bash -lc 'esearch -db sra -query \"PRJNA528413[BioProject]\" | efetch -format runinfo > \"/home/jupyter/Parabricks_pathogen_Test/raw_reads_PRJNA528413/runinfo_PRJNA528413.csv\"'\n",
      "RunInfo saved to: /home/jupyter/Parabricks_pathogen_Test/raw_reads_PRJNA528413/runinfo_PRJNA528413.csv\n",
      "\n",
      "Selected 5 runs from PRJNA528413:\n",
      "   SRR15415520\n",
      "   SRR15462566\n",
      "   SRR15415519\n",
      "   SRR15462117\n",
      "   SRR15441985\n",
      "\n",
      "Downloading FASTQ files into: /home/jupyter/Parabricks_pathogen_Test/raw_reads_PRJNA528413\n",
      "\n",
      "=== Processing run SRR15415520 ===\n",
      "\n",
      "$ prefetch SRR15415520\n",
      "2025-11-16T20:27:54 prefetch.3.2.1: 1) Resolving 'SRR15415520'...\n",
      "2025-11-16T20:27:54 prefetch.3.2.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2025-11-16T20:27:54 prefetch.3.2.1: 1) Downloading 'SRR15415520'...\n",
      "2025-11-16T20:27:54 prefetch.3.2.1:  SRA Normalized Format file is being retrieved\n",
      "2025-11-16T20:27:54 prefetch.3.2.1:  Downloading via HTTPS...\n",
      "2025-11-16T20:31:45 prefetch.3.2.1:  HTTPS download succeed\n",
      "2025-11-16T20:32:06 prefetch.3.2.1:  'SRR15415520' is valid: 10927851219 bytes were streamed from 10927844467\n",
      "2025-11-16T20:32:06 prefetch.3.2.1: 1) 'SRR15415520' was downloaded successfully\n",
      "2025-11-16T20:32:06 prefetch.3.2.1: 1) Resolving 'SRR15415520's dependencies...\n",
      "2025-11-16T20:32:06 prefetch.3.2.1: 'SRR15415520' has 0 unresolved dependencies\n",
      "\n",
      "$ fasterq-dump -e 8 -p -O . SRR15415520\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join   :|-------------------------------------------------- 100%\n",
      "concat :|-------------------------------------------------- 100%\n",
      "spots read      : 163,482\n",
      "reads read      : 163,482\n",
      "reads written   : 163,464\n",
      "reads too short : 18\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "$ gzip -f SRR15415520.fastq\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "import csv\n",
    "\n",
    "# ---------------- CONFIG ----------------\n",
    "PROJECT_ID = \"PRJNA528413\"\n",
    "MAX_RUNS   = 5\n",
    "BASE_DIR   = \"/home/jupyter/Parabricks_pathogen_Test\"\n",
    "OUT_DIR    = os.path.join(BASE_DIR, \"raw_reads_PRJNA528413\")\n",
    "RUNINFO_CSV = os.path.join(OUT_DIR, f\"runinfo_{PROJECT_ID}.csv\")\n",
    "# ----------------------------------------\n",
    "\n",
    "os.makedirs(OUT_DIR, exist_ok=True)\n",
    "\n",
    "def run_cmd(cmd: str):\n",
    "    \"\"\"Run a shell command, raising on error and streaming output.\"\"\"\n",
    "    print(f\"\\n$ {cmd}\")\n",
    "    result = subprocess.run(cmd, shell=True)\n",
    "    if result.returncode != 0:\n",
    "        raise RuntimeError(f\"Command failed with exit code {result.returncode}: {cmd}\")\n",
    "\n",
    "# 1) Ensure Entrez Direct (esearch, efetch) is available\n",
    "if shutil.which(\"esearch\") is None or shutil.which(\"efetch\") is None:\n",
    "    print(\"Entrez Direct not found. Installing via conda (bioconda/entrez-direct)...\")\n",
    "    if shutil.which(\"conda\") is None:\n",
    "        raise RuntimeError(\"conda is not available; cannot auto-install Entrez Direct. \"\n",
    "                           \"Please install entrez-direct manually.\")\n",
    "    run_cmd(\"conda install -y -c bioconda entrez-direct\")\n",
    "\n",
    "# 2) Ensure SRA Toolkit (prefetch, fasterq-dump) is available\n",
    "if shutil.which(\"prefetch\") is None or shutil.which(\"fasterq-dump\") is None:\n",
    "    print(\"SRA Toolkit not found. Installing via conda (bioconda/sra-tools)...\")\n",
    "    if shutil.which(\"conda\") is None:\n",
    "        raise RuntimeError(\"conda is not available; cannot auto-install SRA Toolkit. \"\n",
    "                           \"Please install sra-tools manually.\")\n",
    "    run_cmd(\"conda install -y -c bioconda sra-tools\")\n",
    "\n",
    "# 3) Use Entrez Direct to get RunInfo for this BioProject\n",
    "print(f\"\\nFetching SRA RunInfo for BioProject {PROJECT_ID} ...\")\n",
    "# Note: using bash -lc so the pipe and redirect work correctly\n",
    "run_cmd(\n",
    "    f\"bash -lc 'esearch -db sra -query \\\"{PROJECT_ID}[BioProject]\\\" | \"\n",
    "    f\"efetch -format runinfo > \\\"{RUNINFO_CSV}\\\"'\"\n",
    ")\n",
    "\n",
    "if not os.path.exists(RUNINFO_CSV):\n",
    "    raise FileNotFoundError(f\"RunInfo CSV not created: {RUNINFO_CSV}\")\n",
    "\n",
    "print(f\"RunInfo saved to: {RUNINFO_CSV}\")\n",
    "\n",
    "# 4) Parse RunInfo and select up to MAX_RUNS run accessions\n",
    "runs = []\n",
    "with open(RUNINFO_CSV, newline=\"\") as f:\n",
    "    reader = csv.DictReader(f)\n",
    "    for row in reader:\n",
    "        run_id = row.get(\"Run\")\n",
    "        if run_id:\n",
    "            runs.append(run_id)\n",
    "\n",
    "if not runs:\n",
    "    raise RuntimeError(f\"No runs found in RunInfo for {PROJECT_ID}\")\n",
    "\n",
    "selected_runs = runs[:MAX_RUNS]\n",
    "print(f\"\\nSelected {len(selected_runs)} runs from {PROJECT_ID}:\")\n",
    "for r in selected_runs:\n",
    "    print(\"  \", r)\n",
    "\n",
    "# 5) Download each run with prefetch and fasterq-dump\n",
    "os.chdir(OUT_DIR)\n",
    "print(f\"\\nDownloading FASTQ files into: {OUT_DIR}\")\n",
    "\n",
    "for run_id in selected_runs:\n",
    "    print(f\"\\n=== Processing run {run_id} ===\")\n",
    "\n",
    "    # Download the .sra file into SRA cache\n",
    "    run_cmd(f\"prefetch {run_id}\")\n",
    "\n",
    "    # Convert SRA -> FASTQ (paired if applicable), write into OUT_DIR\n",
    "    # -p = use multiple threads, -e 8 = 8 threads, -O . = output directory\n",
    "    run_cmd(f\"fasterq-dump -e 8 -p -O . {run_id}\")\n",
    "\n",
    "    # Compress outputs for storage efficiency\n",
    "    # Handle both paired-end (_1/_2) and single-end (no suffix)\n",
    "    fq1 = f\"{run_id}_1.fastq\"\n",
    "    fq2 = f\"{run_id}_2.fastq\"\n",
    "    fq_single = f\"{run_id}.fastq\"\n",
    "\n",
    "    if os.path.exists(fq1):\n",
    "        run_cmd(f\"gzip -f {fq1}\")\n",
    "    if os.path.exists(fq2):\n",
    "        run_cmd(f\"gzip -f {fq2}\")\n",
    "    if os.path.exists(fq_single):\n",
    "        run_cmd(f\"gzip -f {fq_single}\")\n",
    "\n",
    "print(\"\\n All selected runs downloaded and compressed.\")\n",
    "print(\"Contents of download directory:\")\n",
    "for fname in sorted(os.listdir(OUT_DIR)):\n",
    "    print(\"  \", fname)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7c0a27ef-9316-4742-b002-817053254114",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyter/Parabricks_pathogen_Test'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4d0076c0-a27c-40ed-ac0d-41bfb37f578f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: pbrun: command not found\n"
     ]
    }
   ],
   "source": [
    "!pbrun minimap2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "09d244fa-b525-4869-ae59-aaa395c9e7dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyter/Parabricks_pathogen_Test'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8fe59659-1493-44b3-9689-fde03392bba1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: pbrun: command not found\n"
     ]
    }
   ],
   "source": [
    "!pbrun --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d4b8623c-5bab-4552-9776-db01fbc7fb25",
   "metadata": {},
   "outputs": [],
   "source": [
    "!find / -name pbrun 2>/dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b0626197-e9df-4e63-b0a9-5a89cc9fe72a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "pbrun: 4.6.0-1\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 pbrun --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "63e4d3a6-1ff1-4e99-8804-12d8144632d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Checking minimap2 test run with 1 sample "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "15dfd980-ff17-4020-a1d2-b26f04ec3890",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gunzip SRR15462566.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4d434c8b-6778-416c-a417-5e9d7bd7b2ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[Parabricks Options Mesg]: Automatically generating ID prefix\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/SRR15462566.fastq\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[PB Info 2025-Nov-19 08:30:07] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:30:07] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 08:30:07] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 08:30:07] ||                                 minimap2                                 ||\n",
      "[PB Info 2025-Nov-19 08:30:07] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:30:10] Reading reference file.\n",
      "[PB Info 2025-Nov-19 08:30:10] -------------------------------------\n",
      "[PB Info 2025-Nov-19 08:30:10] Elapsed-Minutes\tProcessed-Reads\n",
      "[PB Info 2025-Nov-19 08:30:10] -------------------------------------\n",
      "[PB Info 2025-Nov-19 08:30:10] \t0.0\t\t0\n",
      "[PB Info 2025-Nov-19 08:30:16] \t0.1\t\t0\n",
      "[PB Info 2025-Nov-19 08:30:22] \t0.2\t\t0\n",
      "[PB Info 2025-Nov-19 08:30:28] \t0.3\t\t5000\n",
      "[PB Info 2025-Nov-19 08:30:34] \t0.4\t\t45000\n",
      "[PB Info 2025-Nov-19 08:30:40] \t0.5\t\t78466\n",
      "[PB Info 2025-Nov-19 08:30:46] \t0.6\t\t123466\n",
      "[PB Info 2025-Nov-19 08:30:52] \t0.7\t\t163466\n",
      "[PB Info 2025-Nov-19 08:30:55] minimap2 finished in 47.253 seconds\n",
      "[PB Info 2025-Nov-19 08:30:55] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:30:55] ||        Program:                                          minimap2        ||\n",
      "[PB Info 2025-Nov-19 08:30:55] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 08:30:55] ||        Start Time:                       Wed Nov 19 08:30:07 2025        ||\n",
      "[PB Info 2025-Nov-19 08:30:55] ||        End Time:                         Wed Nov 19 08:30:55 2025        ||\n",
      "[PB Info 2025-Nov-19 08:30:55] ||        Total Time:                                     48 seconds        ||\n",
      "[PB Info 2025-Nov-19 08:30:55] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:30:56] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:30:56] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 08:30:56] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 08:30:56] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2025-Nov-19 08:30:56] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:30:56] progressMeter - Percentage\n",
      "[PB Info 2025-Nov-19 08:30:56] 0.0\n",
      "[PB Info 2025-Nov-19 08:31:01] Sorting and Marking: 5.001 seconds\n",
      "[PB Info 2025-Nov-19 08:31:01] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:31:01] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2025-Nov-19 08:31:01] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 08:31:01] ||        Start Time:                       Wed Nov 19 08:30:56 2025        ||\n",
      "[PB Info 2025-Nov-19 08:31:01] ||        End Time:                         Wed Nov 19 08:31:01 2025        ||\n",
      "[PB Info 2025-Nov-19 08:31:01] ||        Total Time:                                      5 seconds        ||\n",
      "[PB Info 2025-Nov-19 08:31:01] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:31:01] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:31:01] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 08:31:01] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 08:31:01] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2025-Nov-19 08:31:01] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:31:02] Using PBBinBamFile for BAM writing\n",
      "[PB Info 2025-Nov-19 08:31:02] progressMeter -\tPercentage\n",
      "[PB Info 2025-Nov-19 08:31:12] 98.8\n",
      "[PB Info 2025-Nov-19 08:31:22] 100.0\n",
      "[PB Info 2025-Nov-19 08:31:22] BQSR and writing final BAM:  20.008 seconds\n",
      "[PB Info 2025-Nov-19 08:31:22] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 08:31:22] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2025-Nov-19 08:31:22] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 08:31:22] ||        Start Time:                       Wed Nov 19 08:31:01 2025        ||\n",
      "[PB Info 2025-Nov-19 08:31:22] ||        End Time:                         Wed Nov 19 08:31:22 2025        ||\n",
      "[PB Info 2025-Nov-19 08:31:22] ||        Total Time:                                     21 seconds        ||\n",
      "[PB Info 2025-Nov-19 08:31:22] ------------------------------------------------------------------------------\n",
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all \\\n",
    "  -v $(pwd):/workdir \\\n",
    "  -w /workdir \\\n",
    "  nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 \\\n",
    "  pbrun minimap2 \\\n",
    "    --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "    --in-fq SRR15462566.fastq \\\n",
    "    --out-bam isolate_01_sorted.bam \\\n",
    "    --preset map-ont"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3d25f1ab-65df-4ffc-834a-42a1d84ba220",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jupyter/Parabricks_pathogen_Test'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "30551a3b-b05f-47b9-831c-1cb62fe425c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Variant calling for haploid bacterial data with ont-germline and inbuilt DeepVariant tool for VCF conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4f59175e-7a81-4837-8aa7-fe24f88d5f30",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "usage: pbrun ont_germline <options>\n",
      "Help: pbrun ont_germline -h\n",
      "pbrun: error: the following arguments are required: --out-bam\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all \\\n",
    "  -v $(pwd):/workdir \\\n",
    "  -w /workdir \\\n",
    "  nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 \\\n",
    "  pbrun ont_germline \\\n",
    "    --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "    --in-bam isolate_01_sorted.bam \\\n",
    "    --out-variants isolate_01_variants.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bfe5c277-287f-460a-928c-5338a81356c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[Parabricks Options Mesg]: Setting --num-streams-per-gpu based on available device memory.\n",
      "[Parabricks Options Mesg]: Please consider using --run-partition for best performance with more than 2 GPUs\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[PB Info 2025-Nov-19 10:11:23] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:11:23] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 10:11:23] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 10:11:23] ||                                 minimap2                                 ||\n",
      "[PB Info 2025-Nov-19 10:11:23] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:11:26] Reading reference file.\n",
      "[PB Info 2025-Nov-19 10:11:26] -------------------------------------\n",
      "[PB Info 2025-Nov-19 10:11:26] Elapsed-Minutes\tProcessed-Reads\n",
      "[PB Info 2025-Nov-19 10:11:26] -------------------------------------\n",
      "[PB Info 2025-Nov-19 10:11:26] \t0.0\t\t0\n",
      "[PB Info 2025-Nov-19 10:11:32] \t0.1\t\t0\n",
      "[PB Info 2025-Nov-19 10:11:38] \t0.2\t\t0\n",
      "[PB Info 2025-Nov-19 10:11:44] \t0.3\t\t0\n",
      "[PB Info 2025-Nov-19 10:11:50] \t0.4\t\t0\n",
      "[PB Info 2025-Nov-19 10:11:56] \t0.5\t\t0\n",
      "[PB Info 2025-Nov-19 10:12:02] \t0.6\t\t5000\n",
      "[PB Info 2025-Nov-19 10:12:08] \t0.7\t\t5000\n",
      "[PB Info 2025-Nov-19 10:12:14] \t0.8\t\t15000\n",
      "[PB Info 2025-Nov-19 10:12:20] \t0.9\t\t25000\n",
      "[PB Info 2025-Nov-19 10:12:26] \t1.0\t\t25000\n",
      "[PB Info 2025-Nov-19 10:12:32] \t1.1\t\t35000\n",
      "[PB Info 2025-Nov-19 10:12:38] \t1.2\t\t35000\n",
      "[PB Info 2025-Nov-19 10:12:44] \t1.3\t\t40000\n",
      "[PB Info 2025-Nov-19 10:12:50] \t1.4\t\t55000\n",
      "[PB Info 2025-Nov-19 10:12:56] \t1.5\t\t65000\n",
      "[PB Info 2025-Nov-19 10:13:02] \t1.6\t\t80000\n",
      "[PB Info 2025-Nov-19 10:13:08] \t1.7\t\t120000\n",
      "[PB Info 2025-Nov-19 10:13:14] \t1.8\t\t125000\n",
      "[PB Info 2025-Nov-19 10:13:20] \t1.9\t\t140000\n",
      "[PB Info 2025-Nov-19 10:13:26] minimap2 finished in 122.977 seconds\n",
      "[PB Info 2025-Nov-19 10:13:26] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:26] ||        Program:                                          minimap2        ||\n",
      "[PB Info 2025-Nov-19 10:13:26] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 10:13:26] ||        Start Time:                       Wed Nov 19 10:11:23 2025        ||\n",
      "[PB Info 2025-Nov-19 10:13:26] ||        End Time:                         Wed Nov 19 10:13:26 2025        ||\n",
      "[PB Info 2025-Nov-19 10:13:26] ||        Total Time:                            2 minutes 3 seconds        ||\n",
      "[PB Info 2025-Nov-19 10:13:26] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:27] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:27] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 10:13:27] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 10:13:27] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2025-Nov-19 10:13:27] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:27] progressMeter - Percentage\n",
      "[PB Info 2025-Nov-19 10:13:27] 0.0\n",
      "[PB Info 2025-Nov-19 10:13:32] Sorting and Marking: 5.000 seconds\n",
      "[PB Info 2025-Nov-19 10:13:32] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:32] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2025-Nov-19 10:13:32] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 10:13:32] ||        Start Time:                       Wed Nov 19 10:13:27 2025        ||\n",
      "[PB Info 2025-Nov-19 10:13:32] ||        End Time:                         Wed Nov 19 10:13:32 2025        ||\n",
      "[PB Info 2025-Nov-19 10:13:32] ||        Total Time:                                      5 seconds        ||\n",
      "[PB Info 2025-Nov-19 10:13:32] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:32] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:32] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 10:13:32] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 10:13:32] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2025-Nov-19 10:13:32] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:33] Using PBBinBamFile for BAM writing\n",
      "[PB Info 2025-Nov-19 10:13:33] progressMeter -\tPercentage\n",
      "[PB Info 2025-Nov-19 10:13:43] 98.7\n",
      "[PB Info 2025-Nov-19 10:13:53] 100.0\n",
      "[PB Info 2025-Nov-19 10:13:53] BQSR and writing final BAM:  20.010 seconds\n",
      "[PB Info 2025-Nov-19 10:13:53] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:53] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2025-Nov-19 10:13:53] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 10:13:53] ||        Start Time:                       Wed Nov 19 10:13:32 2025        ||\n",
      "[PB Info 2025-Nov-19 10:13:53] ||        End Time:                         Wed Nov 19 10:13:53 2025        ||\n",
      "[PB Info 2025-Nov-19 10:13:53] ||        Total Time:                                     21 seconds        ||\n",
      "[PB Info 2025-Nov-19 10:13:53] ------------------------------------------------------------------------------\n",
      "[Parabricks Options Mesg]: Please consider using --run-partition for best performance with more than 2 GPUs\n",
      "Detected 4 CUDA Capable device(s), considering 4 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.9 / 12.9\n",
      "Using model for CUDA Capability Major/Minor version number:    75\n",
      "[PB Info 2025-Nov-19 10:13:58] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:58] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 10:13:58] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 10:13:58] ||                                deepvariant                               ||\n",
      "[PB Info 2025-Nov-19 10:13:58] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 10:13:58] Starting DeepVariant\n",
      "[PB Info 2025-Nov-19 10:13:58] Running with 4 GPU devices, each with 2 group instances and 6 workers\n",
      "[PB Info 2025-Nov-19 10:13:59] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2025-Nov-19 10:14:05] ProgressMeter -\tNC_000913.3:0\t0.1\n",
      "[PB Info 2025-Nov-19 10:14:11] ProgressMeter -\tNC_000913.3:0\t0.2\n",
      "[PB Info 2025-Nov-19 10:14:17] ProgressMeter -\tNC_000913.3:0\t0.3\n",
      "[PB Info 2025-Nov-19 10:14:23] ProgressMeter -\tNC_000913.3:0\t0.4\n",
      "[PB Info 2025-Nov-19 10:14:29] ProgressMeter -\tNC_000913.3:0\t0.5\n",
      "[PB Info 2025-Nov-19 10:14:35] ProgressMeter -\tNC_000913.3:0\t0.6\n",
      "[PB Info 2025-Nov-19 10:14:41] ProgressMeter -\tNC_000913.3:0\t0.7\n",
      "[PB Info 2025-Nov-19 10:14:47] ProgressMeter -\tNC_000913.3:0\t0.8\n",
      "[PB Info 2025-Nov-19 10:14:53] ProgressMeter -\tNC_000913.3:0\t0.9\n",
      "[PB Info 2025-Nov-19 10:14:59] ProgressMeter -\tNC_000913.3:0\t1.0\n",
      "[PB Info 2025-Nov-19 10:15:05] ProgressMeter -\tNC_000913.3:0\t1.1\n",
      "[PB Info 2025-Nov-19 10:15:11] ProgressMeter -\tNC_000913.3:0\t1.2\n",
      "[PB Info 2025-Nov-19 10:15:17] ProgressMeter -\tNC_000913.3:0\t1.3\n",
      "[PB Info 2025-Nov-19 10:15:23] ProgressMeter -\tNC_000913.3:0\t1.4\n",
      "[PB Info 2025-Nov-19 10:15:29] ProgressMeter -\tNC_000913.3:0\t1.5\n",
      "[PB Info 2025-Nov-19 10:15:35] ProgressMeter -\tNC_000913.3:0\t1.6\n",
      "[PB Info 2025-Nov-19 10:15:41] ProgressMeter -\tNC_000913.3:0\t1.7\n",
      "[PB Info 2025-Nov-19 10:15:47] ProgressMeter -\tNC_000913.3:0\t1.8\n",
      "[PB Info 2025-Nov-19 10:15:53] ProgressMeter -\tNC_000913.3:0\t1.9\n",
      "[PB Info 2025-Nov-19 10:15:59] ProgressMeter -\tNC_000913.3:0\t2.0\n",
      "[PB Info 2025-Nov-19 10:16:05] ProgressMeter -\tNC_000913.3:0\t2.1\n",
      "[PB Info 2025-Nov-19 10:16:11] ProgressMeter -\tNC_000913.3:0\t2.2\n",
      "[PB Info 2025-Nov-19 10:16:17] ProgressMeter -\tNC_000913.3:0\t2.3\n",
      "[PB Info 2025-Nov-19 10:16:23] ProgressMeter -\tNC_000913.3:0\t2.4\n",
      "[PB Info 2025-Nov-19 10:16:29] ProgressMeter -\tNC_000913.3:0\t2.5\n",
      "[PB Info 2025-Nov-19 10:16:35] ProgressMeter -\tNC_000913.3:0\t2.6\n",
      "[PB Info 2025-Nov-19 10:16:41] ProgressMeter -\tNC_000913.3:0\t2.7\n",
      "[PB Info 2025-Nov-19 10:16:47] ProgressMeter -\tNC_000913.3:0\t2.8\n",
      "[PB Info 2025-Nov-19 10:16:53] ProgressMeter -\tNC_000913.3:0\t2.9\n",
      "[PB Info 2025-Nov-19 10:16:59] ProgressMeter -\tNC_000913.3:0\t3.0\n",
      "[PB Info 2025-Nov-19 10:17:05] ProgressMeter -\tNC_000913.3:0\t3.1\n",
      "[PB Info 2025-Nov-19 10:17:11] ProgressMeter -\tNC_000913.3:0\t3.2\n",
      "[PB Info 2025-Nov-19 10:17:17] ProgressMeter -\tNC_000913.3:0\t3.3\n",
      "[PB Info 2025-Nov-19 10:17:23] ProgressMeter -\tNC_000913.3:0\t3.4\n",
      "[PB Info 2025-Nov-19 10:17:29] ProgressMeter -\tNC_000913.3:0\t3.5\n",
      "[PB Info 2025-Nov-19 10:17:35] ProgressMeter -\tNC_000913.3:0\t3.6\n",
      "Killed\n",
      "/usr/local/parabricks/binaries/bin/deepsomatic 4 2 --ref /workdir/Ecoli_K12_MG1655_ref.fasta --reads /workdir/isolate_01_processed.bam -o /workdir/isolate_01_variants.vcf -n 6 --model /usr/local/parabricks/binaries/model/75/ont/deepvariant.eng -long_reads --sort_by_haplotypes --add_hp_channel --parse_sam_aux_fields -norealign_reads --vsc_min_fraction_indels 0.12 --track_ref_reads --phase_reads --pileup_image_width 99 --max_reads_per_partition 600 --partition_size 25000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.08 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup diff_channels --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10\n",
      "Process terminated with signal [SIGKILL: 9]. SIGKILL cannot be caught. A common reason for SIGKILL is running out of\n",
      "host memory. If the user has root access, they may be able to check by running `sudo journalctl -k --since \"<#> minutes\n",
      "ago\" | grep \"Killed process\"` to see the reason why processes were recently killed.\n",
      "Process exited with failure. Please check the logs.\n",
      "For technical support visit https://docs.nvidia.com/clara/index.html#parabricks\n",
      "Exiting...\n",
      "\n",
      "\n",
      "Could not run deepvariant as part of ont_germline pipeline\n",
      "Exiting pbrun ...\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all \\\n",
    "  -v $(pwd):/workdir \\\n",
    "  -w /workdir \\\n",
    "  nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 \\\n",
    "  pbrun ont_germline \\\n",
    "    --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "    --in-bam isolate_01_sorted.bam \\\n",
    "    --out-variants isolate_01_variants.vcf \\\n",
    "    --out-bam isolate_01_processed.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "51e317c2-1fa9-4b55-88ab-e6945ed087be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Startimg new "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "757a878c-3ca1-4d80-af52-ac4ba8efbfc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[Parabricks Options Mesg]: Setting --num-streams-per-gpu based on available device memory.\n",
      "[Parabricks Options Mesg]: Please consider using --run-partition for best performance with more than 2 GPUs\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[PB Info 2025-Nov-19 12:24:19] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:24:19] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 12:24:19] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 12:24:19] ||                                 minimap2                                 ||\n",
      "[PB Info 2025-Nov-19 12:24:19] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:24:20] Reading reference file.\n",
      "[PB Info 2025-Nov-19 12:24:20] -------------------------------------\n",
      "[PB Info 2025-Nov-19 12:24:20] Elapsed-Minutes\tProcessed-Reads\n",
      "[PB Info 2025-Nov-19 12:24:20] -------------------------------------\n",
      "[PB Info 2025-Nov-19 12:24:20] \t0.0\t\t0\n",
      "[PB Info 2025-Nov-19 12:24:26] \t0.1\t\t0\n",
      "[PB Info 2025-Nov-19 12:24:32] \t0.2\t\t0\n",
      "[PB Info 2025-Nov-19 12:24:38] \t0.3\t\t0\n",
      "[PB Info 2025-Nov-19 12:24:44] \t0.4\t\t0\n",
      "[PB Info 2025-Nov-19 12:24:50] \t0.5\t\t0\n",
      "[PB Info 2025-Nov-19 12:24:56] \t0.6\t\t0\n",
      "[PB Info 2025-Nov-19 12:25:02] \t0.7\t\t0\n",
      "[PB Info 2025-Nov-19 12:25:08] \t0.8\t\t5000\n",
      "[PB Info 2025-Nov-19 12:25:14] \t0.9\t\t10000\n",
      "[PB Info 2025-Nov-19 12:25:20] \t1.0\t\t15000\n",
      "[PB Info 2025-Nov-19 12:25:26] \t1.1\t\t20000\n",
      "[PB Info 2025-Nov-19 12:25:32] \t1.2\t\t20000\n",
      "[PB Info 2025-Nov-19 12:25:38] \t1.3\t\t25000\n",
      "[PB Info 2025-Nov-19 12:25:44] \t1.4\t\t30000\n",
      "[PB Info 2025-Nov-19 12:25:50] \t1.5\t\t40000\n",
      "[PB Info 2025-Nov-19 12:25:56] \t1.6\t\t40000\n",
      "[PB Info 2025-Nov-19 12:26:02] \t1.7\t\t50000\n",
      "[PB Info 2025-Nov-19 12:26:08] \t1.8\t\t65000\n",
      "[PB Info 2025-Nov-19 12:26:14] \t1.9\t\t85000\n",
      "[PB Info 2025-Nov-19 12:26:20] \t2.0\t\t105000\n",
      "[PB Info 2025-Nov-19 12:26:26] \t2.1\t\t120000\n",
      "[PB Info 2025-Nov-19 12:26:32] \t2.2\t\t135000\n",
      "[PB Info 2025-Nov-19 12:26:38] \t2.3\t\t160000\n",
      "[PB Info 2025-Nov-19 12:26:42] minimap2 finished in 143.208 seconds\n",
      "[PB Info 2025-Nov-19 12:26:42] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:26:42] ||        Program:                                          minimap2        ||\n",
      "[PB Info 2025-Nov-19 12:26:42] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 12:26:42] ||        Start Time:                       Wed Nov 19 12:24:19 2025        ||\n",
      "[PB Info 2025-Nov-19 12:26:42] ||        End Time:                         Wed Nov 19 12:26:42 2025        ||\n",
      "[PB Info 2025-Nov-19 12:26:42] ||        Total Time:                           2 minutes 23 seconds        ||\n",
      "[PB Info 2025-Nov-19 12:26:42] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:26:43] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:26:43] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 12:26:43] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 12:26:43] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2025-Nov-19 12:26:43] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:26:43] progressMeter - Percentage\n",
      "[PB Info 2025-Nov-19 12:26:43] 0.0\n",
      "[PB Info 2025-Nov-19 12:26:48] Sorting and Marking: 5.001 seconds\n",
      "[PB Info 2025-Nov-19 12:26:48] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:26:48] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2025-Nov-19 12:26:48] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 12:26:48] ||        Start Time:                       Wed Nov 19 12:26:43 2025        ||\n",
      "[PB Info 2025-Nov-19 12:26:48] ||        End Time:                         Wed Nov 19 12:26:48 2025        ||\n",
      "[PB Info 2025-Nov-19 12:26:48] ||        Total Time:                                      5 seconds        ||\n",
      "[PB Info 2025-Nov-19 12:26:48] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:26:48] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:26:48] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 12:26:48] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 12:26:48] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2025-Nov-19 12:26:48] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:26:49] Using PBBinBamFile for BAM writing\n",
      "[PB Info 2025-Nov-19 12:26:49] progressMeter -\tPercentage\n",
      "[PB Info 2025-Nov-19 12:26:59] 98.5\n",
      "[PB Info 2025-Nov-19 12:27:09] 100.0\n",
      "[PB Info 2025-Nov-19 12:27:09] BQSR and writing final BAM:  20.018 seconds\n",
      "[PB Info 2025-Nov-19 12:27:09] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:27:09] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2025-Nov-19 12:27:09] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 12:27:09] ||        Start Time:                       Wed Nov 19 12:26:48 2025        ||\n",
      "[PB Info 2025-Nov-19 12:27:09] ||        End Time:                         Wed Nov 19 12:27:09 2025        ||\n",
      "[PB Info 2025-Nov-19 12:27:09] ||        Total Time:                                     21 seconds        ||\n",
      "[PB Info 2025-Nov-19 12:27:09] ------------------------------------------------------------------------------\n",
      "[Parabricks Options Mesg]: Please consider using --run-partition for best performance with more than 2 GPUs\n",
      "Detected 4 CUDA Capable device(s), considering 4 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.9 / 12.9\n",
      "Using model for CUDA Capability Major/Minor version number:    75\n",
      "[PB Info 2025-Nov-19 12:27:13] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:27:13] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 12:27:13] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 12:27:13] ||                                deepvariant                               ||\n",
      "[PB Info 2025-Nov-19 12:27:13] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:27:14] Starting DeepVariant\n",
      "[PB Info 2025-Nov-19 12:27:14] Running with 4 GPU devices, each with 2 group instances and 6 workers\n",
      "[PB Info 2025-Nov-19 12:27:14] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2025-Nov-19 12:27:20] ProgressMeter -\tNC_000913.3:0\t0.1\n",
      "[PB Info 2025-Nov-19 12:27:26] ProgressMeter -\tNC_000913.3:0\t0.2\n",
      "[PB Info 2025-Nov-19 12:27:32] ProgressMeter -\tNC_000913.3:0\t0.3\n",
      "[PB Info 2025-Nov-19 12:27:38] ProgressMeter -\tNC_000913.3:0\t0.4\n",
      "[PB Info 2025-Nov-19 12:27:44] ProgressMeter -\tNC_000913.3:0\t0.5\n",
      "[PB Info 2025-Nov-19 12:27:50] ProgressMeter -\tNC_000913.3:0\t0.6\n",
      "[PB Info 2025-Nov-19 12:27:56] ProgressMeter -\tNC_000913.3:0\t0.7\n",
      "[PB Info 2025-Nov-19 12:28:02] ProgressMeter -\tNC_000913.3:0\t0.8\n",
      "[PB Info 2025-Nov-19 12:28:08] ProgressMeter -\tNC_000913.3:0\t0.9\n",
      "[PB Info 2025-Nov-19 12:28:14] ProgressMeter -\tNC_000913.3:0\t1.0\n",
      "[PB Info 2025-Nov-19 12:28:20] ProgressMeter -\tNC_000913.3:0\t1.1\n",
      "[PB Info 2025-Nov-19 12:28:26] ProgressMeter -\tNC_000913.3:0\t1.2\n",
      "[PB Info 2025-Nov-19 12:28:32] ProgressMeter -\tNC_000913.3:0\t1.3\n",
      "[PB Info 2025-Nov-19 12:28:38] ProgressMeter -\tNC_000913.3:0\t1.4\n",
      "[PB Info 2025-Nov-19 12:28:44] ProgressMeter -\tNC_000913.3:0\t1.5\n",
      "[PB Info 2025-Nov-19 12:28:50] ProgressMeter -\tNC_000913.3:0\t1.6\n",
      "[PB Info 2025-Nov-19 12:28:56] ProgressMeter -\tNC_000913.3:0\t1.7\n",
      "[PB Info 2025-Nov-19 12:29:02] ProgressMeter -\tNC_000913.3:0\t1.8\n",
      "[PB Info 2025-Nov-19 12:29:08] ProgressMeter -\tNC_000913.3:0\t1.9\n",
      "[PB Info 2025-Nov-19 12:29:14] ProgressMeter -\tNC_000913.3:0\t2.0\n",
      "[PB Info 2025-Nov-19 12:29:20] ProgressMeter -\tNC_000913.3:0\t2.1\n",
      "[PB Info 2025-Nov-19 12:29:26] ProgressMeter -\tNC_000913.3:0\t2.2\n",
      "[PB Info 2025-Nov-19 12:29:32] ProgressMeter -\tNC_000913.3:0\t2.3\n",
      "[PB Info 2025-Nov-19 12:29:38] ProgressMeter -\tNC_000913.3:0\t2.4\n",
      "[PB Info 2025-Nov-19 12:29:44] ProgressMeter -\tNC_000913.3:0\t2.5\n",
      "[PB Info 2025-Nov-19 12:29:50] ProgressMeter -\tNC_000913.3:0\t2.6\n",
      "[PB Info 2025-Nov-19 12:29:56] ProgressMeter -\tNC_000913.3:0\t2.7\n",
      "[PB Info 2025-Nov-19 12:30:02] ProgressMeter -\tNC_000913.3:0\t2.8\n",
      "[PB Info 2025-Nov-19 12:30:08] ProgressMeter -\tNC_000913.3:0\t2.9\n",
      "[PB Info 2025-Nov-19 12:30:14] ProgressMeter -\tNC_000913.3:0\t3.0\n",
      "[PB Info 2025-Nov-19 12:30:20] ProgressMeter -\tNC_000913.3:0\t3.1\n",
      "[PB Info 2025-Nov-19 12:30:26] ProgressMeter -\tNC_000913.3:0\t3.2\n",
      "[PB Info 2025-Nov-19 12:30:32] ProgressMeter -\tNC_000913.3:0\t3.3\n",
      "[PB Info 2025-Nov-19 12:30:38] ProgressMeter -\tNC_000913.3:0\t3.4\n",
      "[PB Info 2025-Nov-19 12:30:44] ProgressMeter -\tNC_000913.3:0\t3.5\n",
      "[PB Info 2025-Nov-19 12:30:50] ProgressMeter -\tNC_000913.3:0\t3.6\n",
      "[PB Info 2025-Nov-19 12:30:56] ProgressMeter -\tNC_000913.3:0\t3.7\n",
      "[PB Info 2025-Nov-19 12:31:02] ProgressMeter -\tNC_000913.3:0\t3.8\n",
      "[PB Info 2025-Nov-19 12:31:08] ProgressMeter -\tNC_000913.3:0\t3.9\n",
      "[PB Info 2025-Nov-19 12:31:14] ProgressMeter -\tNC_000913.3:0\t4.0\n",
      "[PB Info 2025-Nov-19 12:31:20] ProgressMeter -\tNC_000913.3:0\t4.1\n",
      "[PB Info 2025-Nov-19 12:31:26] ProgressMeter -\tNC_000913.3:0\t4.2\n",
      "[PB Info 2025-Nov-19 12:31:32] ProgressMeter -\tNC_000913.3:0\t4.3\n",
      "[PB Info 2025-Nov-19 12:31:38] ProgressMeter -\tNC_000913.3:0\t4.4\n",
      "[PB Info 2025-Nov-19 12:31:44] ProgressMeter -\tNC_000913.3:0\t4.5\n",
      "[PB Info 2025-Nov-19 12:31:50] ProgressMeter -\tNC_000913.3:0\t4.6\n",
      "[PB Info 2025-Nov-19 12:31:56] ProgressMeter -\tNC_000913.3:0\t4.7\n",
      "Killed\n",
      "/usr/local/parabricks/binaries/bin/deepsomatic 4 2 --ref /workdir/Ecoli_K12_MG1655_ref.fasta --reads /workdir/isolate_01_processed.bam -o /workdir/isolate_01_variants.vcf -n 6 --model /usr/local/parabricks/binaries/model/75/ont/deepvariant.eng -long_reads --sort_by_haplotypes --add_hp_channel --parse_sam_aux_fields -norealign_reads --vsc_min_fraction_indels 0.12 --track_ref_reads --phase_reads --pileup_image_width 99 --max_reads_per_partition 600 --partition_size 25000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.08 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup diff_channels --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10\n",
      "Process terminated with signal [SIGKILL: 9]. SIGKILL cannot be caught. A common reason for SIGKILL is running out of\n",
      "host memory. If the user has root access, they may be able to check by running `sudo journalctl -k --since \"<#> minutes\n",
      "ago\" | grep \"Killed process\"` to see the reason why processes were recently killed.\n",
      "Process exited with failure. Please check the logs.\n",
      "For technical support visit https://docs.nvidia.com/clara/index.html#parabricks\n",
      "Exiting...\n",
      "\n",
      "\n",
      "Could not run deepvariant as part of ont_germline pipeline\n",
      "Exiting pbrun ...\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all \\\n",
    "  -v $(pwd):/workdir \\\n",
    "  -w /workdir \\\n",
    "  nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 \\\n",
    "  pbrun ont_germline \\\n",
    "    --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "    --in-bam isolate_01_sorted.bam \\\n",
    "    --out-variants isolate_01_variants.vcf \\\n",
    "    --out-bam isolate_01_processed.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f39a10bb-5bb3-4ee7-89c8-e3d21c73279c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Memory-intensive process, reduce reads per partition "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9c7c799b-c982-4be1-91bb-a2c4cecc3e09",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "usage: pbrun ont_germline <options>\n",
      "Help: pbrun ont_germline -h\n",
      "pbrun: error: unrecognized arguments: --extra-flags --max_reads_per_partition 50 --partition_size 10000\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all \\\n",
    "  --ipc=host \\\n",
    "  -v $(pwd):/workdir \\\n",
    "  -w /workdir \\\n",
    "  nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 \\\n",
    "  pbrun ont_germline \\\n",
    "    --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "    --in-bam isolate_01_sorted.bam \\\n",
    "    --out-variants isolate_01_variants.vcf \\\n",
    "    --out-bam isolate_01_processed.bam \\\n",
    "    --extra-flags \"--max_reads_per_partition 50 --partition_size 10000\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "82818921-7938-4b19-b179-8c6b36a782c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[Parabricks Options Mesg]: Setting --num-streams-per-gpu based on available device memory.\n",
      "[Parabricks Options Mesg]: Please consider using --run-partition for best performance with more than 2 GPUs\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[PB Info 2025-Nov-19 12:47:33] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:47:33] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 12:47:33] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 12:47:33] ||                                 minimap2                                 ||\n",
      "[PB Info 2025-Nov-19 12:47:33] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:47:36] Reading reference file.\n",
      "[PB Info 2025-Nov-19 12:47:36] -------------------------------------\n",
      "[PB Info 2025-Nov-19 12:47:36] Elapsed-Minutes\tProcessed-Reads\n",
      "[PB Info 2025-Nov-19 12:47:36] -------------------------------------\n",
      "[PB Info 2025-Nov-19 12:47:36] \t0.0\t\t0\n",
      "[PB Info 2025-Nov-19 12:47:42] \t0.1\t\t0\n",
      "[PB Info 2025-Nov-19 12:47:48] \t0.2\t\t0\n",
      "[PB Info 2025-Nov-19 12:47:54] \t0.3\t\t0\n",
      "[PB Info 2025-Nov-19 12:48:00] \t0.4\t\t0\n",
      "[PB Info 2025-Nov-19 12:48:06] \t0.5\t\t0\n",
      "[PB Info 2025-Nov-19 12:48:12] \t0.6\t\t5000\n",
      "[PB Info 2025-Nov-19 12:48:18] \t0.7\t\t5000\n",
      "[PB Info 2025-Nov-19 12:48:24] \t0.8\t\t15000\n",
      "[PB Info 2025-Nov-19 12:48:30] \t0.9\t\t20000\n",
      "[PB Info 2025-Nov-19 12:48:36] \t1.0\t\t25000\n",
      "[PB Info 2025-Nov-19 12:48:42] \t1.1\t\t35000\n",
      "[PB Info 2025-Nov-19 12:48:48] \t1.2\t\t35000\n",
      "[PB Info 2025-Nov-19 12:48:54] \t1.3\t\t40000\n",
      "[PB Info 2025-Nov-19 12:49:00] \t1.4\t\t50000\n",
      "[PB Info 2025-Nov-19 12:49:06] \t1.5\t\t55000\n",
      "[PB Info 2025-Nov-19 12:49:12] \t1.6\t\t95000\n",
      "[PB Info 2025-Nov-19 12:49:18] \t1.7\t\t110000\n",
      "[PB Info 2025-Nov-19 12:49:24] \t1.8\t\t120000\n",
      "[PB Info 2025-Nov-19 12:49:30] \t1.9\t\t130000\n",
      "[PB Info 2025-Nov-19 12:49:36] \t2.0\t\t159799\n",
      "[PB Info 2025-Nov-19 12:49:38] minimap2 finished in 125.236 seconds\n",
      "[PB Info 2025-Nov-19 12:49:38] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:49:38] ||        Program:                                          minimap2        ||\n",
      "[PB Info 2025-Nov-19 12:49:38] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 12:49:38] ||        Start Time:                       Wed Nov 19 12:47:33 2025        ||\n",
      "[PB Info 2025-Nov-19 12:49:38] ||        End Time:                         Wed Nov 19 12:49:38 2025        ||\n",
      "[PB Info 2025-Nov-19 12:49:38] ||        Total Time:                            2 minutes 5 seconds        ||\n",
      "[PB Info 2025-Nov-19 12:49:38] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:49:39] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:49:39] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 12:49:39] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 12:49:39] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2025-Nov-19 12:49:39] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:49:39] progressMeter - Percentage\n",
      "[PB Info 2025-Nov-19 12:49:39] 0.0\n",
      "[PB Info 2025-Nov-19 12:49:44] Sorting and Marking: 5.001 seconds\n",
      "[PB Info 2025-Nov-19 12:49:44] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:49:44] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2025-Nov-19 12:49:44] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 12:49:44] ||        Start Time:                       Wed Nov 19 12:49:39 2025        ||\n",
      "[PB Info 2025-Nov-19 12:49:44] ||        End Time:                         Wed Nov 19 12:49:44 2025        ||\n",
      "[PB Info 2025-Nov-19 12:49:44] ||        Total Time:                                      5 seconds        ||\n",
      "[PB Info 2025-Nov-19 12:49:44] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:49:44] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:49:44] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 12:49:44] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 12:49:44] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2025-Nov-19 12:49:44] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:49:46] Using PBBinBamFile for BAM writing\n",
      "[PB Info 2025-Nov-19 12:49:46] progressMeter -\tPercentage\n",
      "[PB Info 2025-Nov-19 12:49:56] 98.6\n",
      "[PB Info 2025-Nov-19 12:50:06] 100.0\n",
      "[PB Info 2025-Nov-19 12:50:06] BQSR and writing final BAM:  20.007 seconds\n",
      "[PB Info 2025-Nov-19 12:50:06] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:50:06] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2025-Nov-19 12:50:06] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 12:50:06] ||        Start Time:                       Wed Nov 19 12:49:44 2025        ||\n",
      "[PB Info 2025-Nov-19 12:50:06] ||        End Time:                         Wed Nov 19 12:50:06 2025        ||\n",
      "[PB Info 2025-Nov-19 12:50:06] ||        Total Time:                                     22 seconds        ||\n",
      "[PB Info 2025-Nov-19 12:50:06] ------------------------------------------------------------------------------\n",
      "[Parabricks Options Mesg]: Please consider using --run-partition for best performance with more than 2 GPUs\n",
      "Detected 4 CUDA Capable device(s), considering 4 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.9 / 12.9\n",
      "Using model for CUDA Capability Major/Minor version number:    75\n",
      "[PB Info 2025-Nov-19 12:50:11] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:50:11] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 12:50:11] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 12:50:11] ||                                deepvariant                               ||\n",
      "[PB Info 2025-Nov-19 12:50:11] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 12:50:12] Starting DeepVariant\n",
      "[PB Info 2025-Nov-19 12:50:12] Running with 4 GPU devices, each with 2 group instances and 6 workers\n",
      "[PB Info 2025-Nov-19 12:50:12] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2025-Nov-19 12:50:18] ProgressMeter -\tNC_000913.3:0\t0.1\n",
      "[PB Info 2025-Nov-19 12:50:24] ProgressMeter -\tNC_000913.3:0\t0.2\n",
      "[PB Info 2025-Nov-19 12:50:30] ProgressMeter -\tNC_000913.3:0\t0.3\n",
      "[PB Info 2025-Nov-19 12:50:36] ProgressMeter -\tNC_000913.3:0\t0.4\n",
      "[PB Info 2025-Nov-19 12:50:42] ProgressMeter -\tNC_000913.3:0\t0.5\n",
      "[PB Info 2025-Nov-19 12:50:48] ProgressMeter -\tNC_000913.3:0\t0.6\n",
      "[PB Info 2025-Nov-19 12:50:54] ProgressMeter -\tNC_000913.3:0\t0.7\n",
      "[PB Info 2025-Nov-19 12:51:00] ProgressMeter -\tNC_000913.3:0\t0.8\n",
      "[PB Info 2025-Nov-19 12:51:06] ProgressMeter -\tNC_000913.3:0\t0.9\n",
      "[PB Info 2025-Nov-19 12:51:12] ProgressMeter -\tNC_000913.3:0\t1.0\n",
      "[PB Info 2025-Nov-19 12:51:18] ProgressMeter -\tNC_000913.3:0\t1.1\n",
      "[PB Info 2025-Nov-19 12:51:24] ProgressMeter -\tNC_000913.3:0\t1.2\n",
      "[PB Info 2025-Nov-19 12:51:30] ProgressMeter -\tNC_000913.3:0\t1.3\n",
      "[PB Info 2025-Nov-19 12:51:36] ProgressMeter -\tNC_000913.3:0\t1.4\n",
      "[PB Info 2025-Nov-19 12:51:42] ProgressMeter -\tNC_000913.3:0\t1.5\n",
      "[PB Info 2025-Nov-19 12:51:48] ProgressMeter -\tNC_000913.3:0\t1.6\n",
      "[PB Info 2025-Nov-19 12:51:54] ProgressMeter -\tNC_000913.3:0\t1.7\n",
      "[PB Info 2025-Nov-19 12:52:00] ProgressMeter -\tNC_000913.3:0\t1.8\n",
      "[PB Info 2025-Nov-19 12:52:06] ProgressMeter -\tNC_000913.3:0\t1.9\n",
      "[PB Info 2025-Nov-19 12:52:12] ProgressMeter -\tNC_000913.3:0\t2.0\n",
      "[PB Info 2025-Nov-19 12:52:18] ProgressMeter -\tNC_000913.3:0\t2.1\n",
      "[PB Info 2025-Nov-19 12:52:24] ProgressMeter -\tNC_000913.3:0\t2.2\n",
      "[PB Info 2025-Nov-19 12:52:30] ProgressMeter -\tNC_000913.3:0\t2.3\n",
      "[PB Info 2025-Nov-19 12:52:36] ProgressMeter -\tNC_000913.3:0\t2.4\n",
      "[PB Info 2025-Nov-19 12:52:42] ProgressMeter -\tNC_000913.3:0\t2.5\n",
      "[PB Info 2025-Nov-19 12:52:48] ProgressMeter -\tNC_000913.3:0\t2.6\n",
      "[PB Info 2025-Nov-19 12:52:54] ProgressMeter -\tNC_000913.3:0\t2.7\n",
      "[PB Info 2025-Nov-19 12:53:00] ProgressMeter -\tNC_000913.3:0\t2.8\n",
      "[PB Info 2025-Nov-19 12:53:06] ProgressMeter -\tNC_000913.3:0\t2.9\n",
      "[PB Info 2025-Nov-19 12:53:12] ProgressMeter -\tNC_000913.3:0\t3.0\n",
      "[PB Info 2025-Nov-19 12:53:18] ProgressMeter -\tNC_000913.3:0\t3.1\n",
      "[PB Info 2025-Nov-19 12:53:24] ProgressMeter -\tNC_000913.3:0\t3.2\n",
      "[PB Info 2025-Nov-19 12:53:30] ProgressMeter -\tNC_000913.3:0\t3.3\n",
      "[PB Info 2025-Nov-19 12:53:36] ProgressMeter -\tNC_000913.3:0\t3.4\n",
      "[PB Info 2025-Nov-19 12:53:42] ProgressMeter -\tNC_000913.3:0\t3.5\n",
      "[PB Info 2025-Nov-19 12:53:48] ProgressMeter -\tNC_000913.3:0\t3.6\n",
      "[PB Info 2025-Nov-19 12:53:54] ProgressMeter -\tNC_000913.3:0\t3.7\n",
      "[PB Info 2025-Nov-19 12:54:00] ProgressMeter -\tNC_000913.3:0\t3.8\n",
      "[PB Info 2025-Nov-19 12:54:06] ProgressMeter -\tNC_000913.3:0\t3.9\n",
      "[PB Info 2025-Nov-19 12:54:12] ProgressMeter -\tNC_000913.3:0\t4.0\n",
      "[PB Info 2025-Nov-19 12:54:18] ProgressMeter -\tNC_000913.3:0\t4.1\n",
      "[PB Info 2025-Nov-19 12:54:24] ProgressMeter -\tNC_000913.3:0\t4.2\n",
      "[PB Info 2025-Nov-19 12:54:30] ProgressMeter -\tNC_000913.3:0\t4.3\n",
      "[PB Info 2025-Nov-19 12:54:36] ProgressMeter -\tNC_000913.3:0\t4.4\n",
      "[PB Info 2025-Nov-19 12:54:42] ProgressMeter -\tNC_000913.3:0\t4.5\n",
      "[PB Info 2025-Nov-19 12:54:48] ProgressMeter -\tNC_000913.3:0\t4.6\n",
      "[PB Info 2025-Nov-19 12:54:54] ProgressMeter -\tNC_000913.3:0\t4.7\n",
      "[PB Info 2025-Nov-19 12:55:00] ProgressMeter -\tNC_000913.3:0\t4.8\n",
      "[PB Info 2025-Nov-19 12:55:06] ProgressMeter -\tNC_000913.3:0\t4.9\n",
      "[PB Info 2025-Nov-19 12:55:12] ProgressMeter -\tNC_000913.3:0\t5.0\n",
      "[PB Info 2025-Nov-19 12:55:18] ProgressMeter -\tNC_000913.3:0\t5.1\n",
      "[PB Info 2025-Nov-19 12:55:24] ProgressMeter -\tNC_000913.3:0\t5.2\n",
      "[PB Info 2025-Nov-19 12:55:30] ProgressMeter -\tNC_000913.3:0\t5.3\n",
      "[PB Info 2025-Nov-19 12:55:36] ProgressMeter -\tNC_000913.3:0\t5.4\n",
      "[PB Info 2025-Nov-19 12:55:42] ProgressMeter -\tNC_000913.3:0\t5.5\n",
      "[PB Info 2025-Nov-19 12:55:48] ProgressMeter -\tNC_000913.3:0\t5.6\n",
      "[PB Info 2025-Nov-19 12:55:54] ProgressMeter -\tNC_000913.3:0\t5.7\n",
      "[PB Info 2025-Nov-19 12:56:00] ProgressMeter -\tNC_000913.3:0\t5.8\n",
      "[PB Info 2025-Nov-19 12:56:06] ProgressMeter -\tNC_000913.3:0\t5.9\n",
      "[PB Info 2025-Nov-19 12:56:12] ProgressMeter -\tNC_000913.3:0\t6.0\n",
      "[PB Info 2025-Nov-19 12:56:18] ProgressMeter -\tNC_000913.3:0\t6.1\n",
      "[PB Info 2025-Nov-19 12:56:24] ProgressMeter -\tNC_000913.3:0\t6.2\n",
      "Killed\n",
      "/usr/local/parabricks/binaries/bin/deepsomatic 4 2 --ref /workdir/Ecoli_K12_MG1655_ref.fasta --reads /workdir/isolate_01_processed.bam -o /workdir/isolate_01_variants.vcf -n 6 --model /usr/local/parabricks/binaries/model/75/ont/deepvariant.eng -long_reads --sort_by_haplotypes --add_hp_channel --parse_sam_aux_fields -norealign_reads --vsc_min_fraction_indels 0.12 --track_ref_reads --phase_reads --pileup_image_width 99 --max_reads_per_partition 600 --partition_size 25000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.08 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup diff_channels --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10\n",
      "Process terminated with signal [SIGKILL: 9]. SIGKILL cannot be caught. A common reason for SIGKILL is running out of\n",
      "host memory. If the user has root access, they may be able to check by running `sudo journalctl -k --since \"<#> minutes\n",
      "ago\" | grep \"Killed process\"` to see the reason why processes were recently killed.\n",
      "Process exited with failure. Please check the logs.\n",
      "For technical support visit https://docs.nvidia.com/clara/index.html#parabricks\n",
      "Exiting...\n",
      "\n",
      "\n",
      "Could not run deepvariant as part of ont_germline pipeline\n",
      "Exiting pbrun ...\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all \\\n",
    "  -v $(pwd):/workdir \\\n",
    "  -w /workdir \\\n",
    "  nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 \\\n",
    "  pbrun ont_germline \\\n",
    "    --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "    --in-bam isolate_01_sorted.bam \\\n",
    "    --out-variants isolate_01_variants.vcf \\\n",
    "    --out-bam isolate_01_processed.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "de556bc6-0639-49f0-9e25-b959604d2a39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nov 19 12:56:29 ac-genomics kernel: Out of memory: Killed process 29280 (deepsomatic) total-vm:531608348kB, anon-rss:421559076kB, file-rss:268736kB, shmem-rss:580288kB, UID:0 pgtables:893636kB oom_score_adj:0\n"
     ]
    }
   ],
   "source": [
    "!sudo journalctl -k --since \"30 minutes ago\" | grep \"Killed process\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aef24359-dc90-45b7-bd70-9a473191e4ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nov 19 12:56:29 ac-genomics kernel: Out of memory: Killed process 29280 (deepsomatic) total-vm:531608348kB, anon-rss:421559076kB, file-rss:268736kB, shmem-rss:580288kB, UID:0 pgtables:893636kB oom_score_adj:0\n"
     ]
    }
   ],
   "source": [
    "!sudo journalctl -k --since \"1 hour ago\" | grep \"Out of memory\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "10197dc1-8576-4bd9-a005-8f37d990f609",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[Parabricks Options Mesg]: Setting --num-streams-per-gpu based on available device memory.\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[PB Info 2025-Nov-19 13:13:30] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:13:30] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 13:13:30] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 13:13:30] ||                                 minimap2                                 ||\n",
      "[PB Info 2025-Nov-19 13:13:30] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:13:32] Reading reference file.\n",
      "[PB Info 2025-Nov-19 13:13:32] -------------------------------------\n",
      "[PB Info 2025-Nov-19 13:13:32] Elapsed-Minutes\tProcessed-Reads\n",
      "[PB Info 2025-Nov-19 13:13:32] -------------------------------------\n",
      "[PB Info 2025-Nov-19 13:13:32] \t0.0\t\t0\n",
      "[PB Info 2025-Nov-19 13:13:38] \t0.1\t\t0\n",
      "[PB Info 2025-Nov-19 13:13:44] \t0.2\t\t0\n",
      "[PB Info 2025-Nov-19 13:13:50] \t0.3\t\t0\n",
      "[PB Info 2025-Nov-19 13:13:56] \t0.4\t\t0\n",
      "[PB Info 2025-Nov-19 13:14:02] \t0.5\t\t0\n",
      "[PB Info 2025-Nov-19 13:14:08] \t0.6\t\t5000\n",
      "[PB Info 2025-Nov-19 13:14:14] \t0.7\t\t10000\n",
      "[PB Info 2025-Nov-19 13:14:20] \t0.8\t\t15000\n",
      "[PB Info 2025-Nov-19 13:14:26] \t0.9\t\t20000\n",
      "[PB Info 2025-Nov-19 13:14:32] \t1.0\t\t30000\n",
      "[PB Info 2025-Nov-19 13:14:38] \t1.1\t\t30000\n",
      "[PB Info 2025-Nov-19 13:14:44] \t1.2\t\t35000\n",
      "[PB Info 2025-Nov-19 13:14:50] \t1.3\t\t40000\n",
      "[PB Info 2025-Nov-19 13:14:56] \t1.4\t\t50000\n",
      "[PB Info 2025-Nov-19 13:15:02] \t1.5\t\t65000\n",
      "[PB Info 2025-Nov-19 13:15:08] \t1.6\t\t75000\n",
      "[PB Info 2025-Nov-19 13:15:14] \t1.7\t\t105000\n",
      "[PB Info 2025-Nov-19 13:15:20] \t1.8\t\t120000\n",
      "[PB Info 2025-Nov-19 13:15:26] \t1.9\t\t130000\n",
      "[PB Info 2025-Nov-19 13:15:32] \t2.0\t\t154799\n",
      "[PB Info 2025-Nov-19 13:15:35] minimap2 finished in 125.074 seconds\n",
      "[PB Info 2025-Nov-19 13:15:35] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:15:35] ||        Program:                                          minimap2        ||\n",
      "[PB Info 2025-Nov-19 13:15:35] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 13:15:35] ||        Start Time:                       Wed Nov 19 13:13:30 2025        ||\n",
      "[PB Info 2025-Nov-19 13:15:35] ||        End Time:                         Wed Nov 19 13:15:35 2025        ||\n",
      "[PB Info 2025-Nov-19 13:15:35] ||        Total Time:                            2 minutes 5 seconds        ||\n",
      "[PB Info 2025-Nov-19 13:15:35] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:15:36] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:15:36] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 13:15:36] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 13:15:36] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2025-Nov-19 13:15:36] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:15:36] progressMeter - Percentage\n",
      "[PB Info 2025-Nov-19 13:15:36] 0.0\n",
      "[PB Info 2025-Nov-19 13:15:41] Sorting and Marking: 5.001 seconds\n",
      "[PB Info 2025-Nov-19 13:15:41] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:15:41] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2025-Nov-19 13:15:41] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 13:15:41] ||        Start Time:                       Wed Nov 19 13:15:36 2025        ||\n",
      "[PB Info 2025-Nov-19 13:15:41] ||        End Time:                         Wed Nov 19 13:15:41 2025        ||\n",
      "[PB Info 2025-Nov-19 13:15:41] ||        Total Time:                                      5 seconds        ||\n",
      "[PB Info 2025-Nov-19 13:15:41] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:15:41] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:15:41] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 13:15:41] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 13:15:41] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2025-Nov-19 13:15:41] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:15:42] Using PBBinBamFile for BAM writing\n",
      "[PB Info 2025-Nov-19 13:15:42] progressMeter -\tPercentage\n",
      "[PB Info 2025-Nov-19 13:15:52] 98.8\n",
      "[PB Info 2025-Nov-19 13:16:02] 100.0\n",
      "[PB Info 2025-Nov-19 13:16:02] BQSR and writing final BAM:  20.009 seconds\n",
      "[PB Info 2025-Nov-19 13:16:02] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:16:02] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2025-Nov-19 13:16:02] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 13:16:02] ||        Start Time:                       Wed Nov 19 13:15:41 2025        ||\n",
      "[PB Info 2025-Nov-19 13:16:02] ||        End Time:                         Wed Nov 19 13:16:02 2025        ||\n",
      "[PB Info 2025-Nov-19 13:16:02] ||        Total Time:                                     21 seconds        ||\n",
      "[PB Info 2025-Nov-19 13:16:02] ------------------------------------------------------------------------------\n",
      "Detected 4 CUDA Capable device(s), considering 4 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.9 / 12.9\n",
      "Using model for CUDA Capability Major/Minor version number:    75\n",
      "[PB Info 2025-Nov-19 13:16:07] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:16:07] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 13:16:07] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 13:16:07] ||                                deepvariant                               ||\n",
      "[PB Info 2025-Nov-19 13:16:07] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:16:07] Starting DeepVariant\n",
      "[PB Info 2025-Nov-19 13:16:07] Running with 2 GPU devices, each with 2 group instances and 6 workers\n",
      "[PB Info 2025-Nov-19 13:16:07] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2025-Nov-19 13:16:13] ProgressMeter -\tNC_000913.3:0\t0.1\n",
      "[PB Info 2025-Nov-19 13:16:19] ProgressMeter -\tNC_000913.3:0\t0.2\n",
      "[PB Info 2025-Nov-19 13:16:25] ProgressMeter -\tNC_000913.3:0\t0.3\n",
      "[PB Info 2025-Nov-19 13:16:31] ProgressMeter -\tNC_000913.3:0\t0.4\n",
      "[PB Info 2025-Nov-19 13:16:37] ProgressMeter -\tNC_000913.3:0\t0.5\n",
      "[PB Info 2025-Nov-19 13:16:43] ProgressMeter -\tNC_000913.3:0\t0.6\n",
      "[PB Info 2025-Nov-19 13:16:49] ProgressMeter -\tNC_000913.3:0\t0.7\n",
      "[PB Info 2025-Nov-19 13:16:55] ProgressMeter -\tNC_000913.3:0\t0.8\n",
      "[PB Info 2025-Nov-19 13:17:01] ProgressMeter -\tNC_000913.3:0\t0.9\n",
      "[PB Info 2025-Nov-19 13:17:07] ProgressMeter -\tNC_000913.3:0\t1.0\n",
      "[PB Info 2025-Nov-19 13:17:13] ProgressMeter -\tNC_000913.3:0\t1.1\n",
      "[PB Info 2025-Nov-19 13:17:19] ProgressMeter -\tNC_000913.3:0\t1.2\n",
      "[PB Info 2025-Nov-19 13:17:25] ProgressMeter -\tNC_000913.3:0\t1.3\n",
      "[PB Info 2025-Nov-19 13:17:31] ProgressMeter -\tNC_000913.3:0\t1.4\n",
      "[PB Info 2025-Nov-19 13:17:37] ProgressMeter -\tNC_000913.3:0\t1.5\n",
      "[PB Info 2025-Nov-19 13:17:43] ProgressMeter -\tNC_000913.3:0\t1.6\n",
      "[PB Info 2025-Nov-19 13:17:49] ProgressMeter -\tNC_000913.3:0\t1.7\n",
      "[PB Info 2025-Nov-19 13:17:55] ProgressMeter -\tNC_000913.3:0\t1.8\n",
      "[PB Info 2025-Nov-19 13:18:01] ProgressMeter -\tNC_000913.3:0\t1.9\n",
      "[PB Info 2025-Nov-19 13:18:07] ProgressMeter -\tNC_000913.3:0\t2.0\n",
      "[PB Info 2025-Nov-19 13:18:13] ProgressMeter -\tNC_000913.3:0\t2.1\n",
      "[PB Info 2025-Nov-19 13:18:19] ProgressMeter -\tNC_000913.3:0\t2.2\n",
      "[PB Info 2025-Nov-19 13:18:25] ProgressMeter -\tNC_000913.3:0\t2.3\n",
      "[PB Info 2025-Nov-19 13:18:31] ProgressMeter -\tNC_000913.3:0\t2.4\n",
      "[PB Info 2025-Nov-19 13:18:37] ProgressMeter -\tNC_000913.3:0\t2.5\n",
      "[PB Info 2025-Nov-19 13:18:43] ProgressMeter -\tNC_000913.3:0\t2.6\n",
      "[PB Info 2025-Nov-19 13:18:49] ProgressMeter -\tNC_000913.3:0\t2.7\n",
      "[PB Info 2025-Nov-19 13:18:55] ProgressMeter -\tNC_000913.3:0\t2.8\n",
      "[PB Info 2025-Nov-19 13:19:01] ProgressMeter -\tNC_000913.3:0\t2.9\n",
      "[PB Info 2025-Nov-19 13:19:07] ProgressMeter -\tNC_000913.3:0\t3.0\n",
      "[PB Info 2025-Nov-19 13:19:13] ProgressMeter -\tNC_000913.3:0\t3.1\n",
      "[PB Info 2025-Nov-19 13:19:19] ProgressMeter -\tNC_000913.3:0\t3.2\n",
      "[PB Info 2025-Nov-19 13:19:25] ProgressMeter -\tNC_000913.3:0\t3.3\n",
      "[PB Info 2025-Nov-19 13:19:31] ProgressMeter -\tNC_000913.3:0\t3.4\n",
      "[PB Info 2025-Nov-19 13:19:37] ProgressMeter -\tNC_000913.3:0\t3.5\n",
      "[PB Info 2025-Nov-19 13:19:43] ProgressMeter -\tNC_000913.3:0\t3.6\n",
      "[PB Info 2025-Nov-19 13:19:49] ProgressMeter -\tNC_000913.3:0\t3.7\n",
      "[PB Info 2025-Nov-19 13:19:55] ProgressMeter -\tNC_000913.3:0\t3.8\n",
      "[PB Info 2025-Nov-19 13:20:01] ProgressMeter -\tNC_000913.3:0\t3.9\n",
      "[PB Info 2025-Nov-19 13:20:07] ProgressMeter -\tNC_000913.3:0\t4.0\n",
      "[PB Info 2025-Nov-19 13:20:13] ProgressMeter -\tNC_000913.3:0\t4.1\n",
      "[PB Info 2025-Nov-19 13:20:19] ProgressMeter -\tNC_000913.3:0\t4.2\n",
      "[PB Info 2025-Nov-19 13:20:25] ProgressMeter -\tNC_000913.3:0\t4.3\n",
      "[PB Info 2025-Nov-19 13:20:31] ProgressMeter -\tNC_000913.3:0\t4.4\n",
      "[PB Info 2025-Nov-19 13:20:37] ProgressMeter -\tNC_000913.3:0\t4.5\n",
      "[PB Info 2025-Nov-19 13:20:43] ProgressMeter -\tNC_000913.3:0\t4.6\n",
      "[PB Info 2025-Nov-19 13:20:49] ProgressMeter -\tNC_000913.3:0\t4.7\n",
      "[PB Info 2025-Nov-19 13:20:55] ProgressMeter -\tNC_000913.3:0\t4.8\n",
      "[PB Info 2025-Nov-19 13:21:01] ProgressMeter -\tNC_000913.3:0\t4.9\n",
      "[PB Info 2025-Nov-19 13:21:07] ProgressMeter -\tNC_000913.3:0\t5.0\n",
      "[PB Info 2025-Nov-19 13:21:13] ProgressMeter -\tNC_000913.3:0\t5.1\n",
      "[PB Info 2025-Nov-19 13:21:19] ProgressMeter -\tNC_000913.3:0\t5.2\n",
      "[PB Info 2025-Nov-19 13:21:25] ProgressMeter -\tNC_000913.3:0\t5.3\n",
      "[PB Info 2025-Nov-19 13:21:36] ProgressMeter -\tNC_000913.3:0\t5.4\n",
      "Killed\n",
      "CUDA_VISIBLE_DEVICES=0,1 /usr/local/parabricks/binaries/bin/deepsomatic 2 2 --ref /workdir/Ecoli_K12_MG1655_ref.fasta --reads /workdir/isolate_01_processed.bam -o 0.vcf  -L \"NC_000913.3\" -n 6 --model /usr/local/parabricks/binaries/model/75/ont/deepvariant.eng -long_reads --sort_by_haplotypes --add_hp_channel --parse_sam_aux_fields -norealign_reads --vsc_min_fraction_indels 0.12 --track_ref_reads --phase_reads --pileup_image_width 99 --max_reads_per_partition 600 --partition_size 25000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.08 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup diff_channels --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10 \n",
      "Process terminated with signal [SIGKILL: 9]. SIGKILL cannot be caught. A common reason for SIGKILL is running out of\n",
      "host memory. If the user has root access, they may be able to check by running `sudo journalctl -k --since \"<#> minutes\n",
      "ago\" | grep \"Killed process\"` to see the reason why processes were recently killed.\n",
      "Process exited with failure. Please check the logs.\n",
      "For technical support visit https://docs.nvidia.com/clara/index.html#parabricks\n",
      "Exiting...\n",
      "\n",
      "\n",
      "Could not run deepvariant as part of ont_germline pipeline\n",
      "Exiting pbrun ...\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all \\\n",
    "  --ipc=host \\\n",
    "  -v $(pwd):/workdir \\\n",
    "  -w /workdir \\\n",
    "  nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 \\\n",
    "  pbrun ont_germline \\\n",
    "    --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "    --in-bam isolate_01_sorted.bam \\\n",
    "    --out-variants isolate_01_variants.vcf \\\n",
    "    --out-bam isolate_01_processed.bam \\\n",
    "    --low-memory \\\n",
    "    --run-partition\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c9c1e1af-71fb-4d44-a165-cb0dcaab6efd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[Parabricks Options Mesg]: Setting --num-streams-per-gpu based on available device memory.\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[PB Info 2025-Nov-19 13:25:20] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:25:20] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 13:25:20] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 13:25:20] ||                                 minimap2                                 ||\n",
      "[PB Info 2025-Nov-19 13:25:20] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:25:22] Reading reference file.\n",
      "[PB Info 2025-Nov-19 13:25:23] -------------------------------------\n",
      "[PB Info 2025-Nov-19 13:25:23] Elapsed-Minutes\tProcessed-Reads\n",
      "[PB Info 2025-Nov-19 13:25:23] -------------------------------------\n",
      "[PB Info 2025-Nov-19 13:25:23] \t0.0\t\t0\n",
      "[PB Info 2025-Nov-19 13:25:29] \t0.1\t\t0\n",
      "[PB Info 2025-Nov-19 13:25:35] \t0.2\t\t0\n",
      "[PB Info 2025-Nov-19 13:25:41] \t0.3\t\t0\n",
      "[PB Info 2025-Nov-19 13:25:47] \t0.4\t\t0\n",
      "[PB Info 2025-Nov-19 13:25:53] \t0.5\t\t0\n",
      "[PB Info 2025-Nov-19 13:25:59] \t0.6\t\t5000\n",
      "[PB Info 2025-Nov-19 13:26:05] \t0.7\t\t5000\n",
      "[PB Info 2025-Nov-19 13:26:11] \t0.8\t\t15000\n",
      "[PB Info 2025-Nov-19 13:26:17] \t0.9\t\t25000\n",
      "[PB Info 2025-Nov-19 13:26:23] \t1.0\t\t30000\n",
      "[PB Info 2025-Nov-19 13:26:29] \t1.1\t\t35000\n",
      "[PB Info 2025-Nov-19 13:26:35] \t1.2\t\t35000\n",
      "[PB Info 2025-Nov-19 13:26:41] \t1.3\t\t45000\n",
      "[PB Info 2025-Nov-19 13:26:47] \t1.4\t\t55000\n",
      "[PB Info 2025-Nov-19 13:26:53] \t1.5\t\t75000\n",
      "[PB Info 2025-Nov-19 13:26:59] \t1.6\t\t100000\n",
      "[PB Info 2025-Nov-19 13:27:05] \t1.7\t\t120000\n",
      "[PB Info 2025-Nov-19 13:27:11] \t1.8\t\t130000\n",
      "[PB Info 2025-Nov-19 13:27:17] \t1.9\t\t155000\n",
      "[PB Info 2025-Nov-19 13:27:20] minimap2 finished in 119.680 seconds\n",
      "[PB Info 2025-Nov-19 13:27:20] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:20] ||        Program:                                          minimap2        ||\n",
      "[PB Info 2025-Nov-19 13:27:20] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 13:27:20] ||        Start Time:                       Wed Nov 19 13:25:20 2025        ||\n",
      "[PB Info 2025-Nov-19 13:27:20] ||        End Time:                         Wed Nov 19 13:27:20 2025        ||\n",
      "[PB Info 2025-Nov-19 13:27:20] ||        Total Time:                            2 minutes 0 seconds        ||\n",
      "[PB Info 2025-Nov-19 13:27:20] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:21] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:21] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 13:27:21] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 13:27:21] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2025-Nov-19 13:27:21] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:21] progressMeter - Percentage\n",
      "[PB Info 2025-Nov-19 13:27:21] 0.0\n",
      "[PB Info 2025-Nov-19 13:27:26] Sorting and Marking: 5.002 seconds\n",
      "[PB Info 2025-Nov-19 13:27:26] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:26] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2025-Nov-19 13:27:26] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 13:27:26] ||        Start Time:                       Wed Nov 19 13:27:21 2025        ||\n",
      "[PB Info 2025-Nov-19 13:27:26] ||        End Time:                         Wed Nov 19 13:27:26 2025        ||\n",
      "[PB Info 2025-Nov-19 13:27:26] ||        Total Time:                                      5 seconds        ||\n",
      "[PB Info 2025-Nov-19 13:27:26] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:26] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:26] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 13:27:26] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 13:27:26] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2025-Nov-19 13:27:26] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:27] Using PBBinBamFile for BAM writing\n",
      "[PB Info 2025-Nov-19 13:27:27] progressMeter -\tPercentage\n",
      "[PB Info 2025-Nov-19 13:27:37] 98.7\n",
      "[PB Info 2025-Nov-19 13:27:47] 100.0\n",
      "[PB Info 2025-Nov-19 13:27:47] BQSR and writing final BAM:  20.009 seconds\n",
      "[PB Info 2025-Nov-19 13:27:47] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:47] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2025-Nov-19 13:27:47] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 13:27:47] ||        Start Time:                       Wed Nov 19 13:27:26 2025        ||\n",
      "[PB Info 2025-Nov-19 13:27:47] ||        End Time:                         Wed Nov 19 13:27:47 2025        ||\n",
      "[PB Info 2025-Nov-19 13:27:47] ||        Total Time:                                     21 seconds        ||\n",
      "[PB Info 2025-Nov-19 13:27:47] ------------------------------------------------------------------------------\n",
      "Detected 4 CUDA Capable device(s), considering 4 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.9 / 12.9\n",
      "Using model for CUDA Capability Major/Minor version number:    75\n",
      "[PB Info 2025-Nov-19 13:27:52] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:52] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 13:27:52] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 13:27:52] ||                                deepvariant                               ||\n",
      "[PB Info 2025-Nov-19 13:27:52] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 13:27:53] Starting DeepVariant\n",
      "[PB Info 2025-Nov-19 13:27:53] Running with 2 GPU devices, each with 2 group instances and 6 workers\n",
      "[PB Info 2025-Nov-19 13:27:53] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2025-Nov-19 13:27:59] ProgressMeter -\tNC_000913.3:0\t0.1\n",
      "[PB Info 2025-Nov-19 13:28:05] ProgressMeter -\tNC_000913.3:0\t0.2\n",
      "[PB Info 2025-Nov-19 13:28:11] ProgressMeter -\tNC_000913.3:0\t0.3\n",
      "[PB Info 2025-Nov-19 13:28:17] ProgressMeter -\tNC_000913.3:0\t0.4\n",
      "[PB Info 2025-Nov-19 13:28:23] ProgressMeter -\tNC_000913.3:0\t0.5\n",
      "[PB Info 2025-Nov-19 13:28:29] ProgressMeter -\tNC_000913.3:0\t0.6\n",
      "[PB Info 2025-Nov-19 13:28:35] ProgressMeter -\tNC_000913.3:0\t0.7\n",
      "[PB Info 2025-Nov-19 13:28:41] ProgressMeter -\tNC_000913.3:0\t0.8\n",
      "Killed\n",
      "CUDA_VISIBLE_DEVICES=0,1 /usr/local/parabricks/binaries/bin/deepsomatic 2 2 --ref /workdir/Ecoli_K12_MG1655_ref.fasta --reads /workdir/isolate_01_processed.bam -o 0.vcf  -L \"NC_000913.3\" -n 6 --model /usr/local/parabricks/binaries/model/75/ont/deepvariant.eng -long_reads --sort_by_haplotypes --add_hp_channel --parse_sam_aux_fields -norealign_reads --vsc_min_fraction_indels 0.12 --track_ref_reads --phase_reads --pileup_image_width 99 --max_reads_per_partition 600 --partition_size 25000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.08 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup diff_channels --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10 \n",
      "Process terminated with signal [SIGKILL: 9]. SIGKILL cannot be caught. A common reason for SIGKILL is running out of\n",
      "host memory. If the user has root access, they may be able to check by running `sudo journalctl -k --since \"<#> minutes\n",
      "ago\" | grep \"Killed process\"` to see the reason why processes were recently killed.\n",
      "Process exited with failure. Please check the logs.\n",
      "For technical support visit https://docs.nvidia.com/clara/index.html#parabricks\n",
      "Exiting...\n",
      "\n",
      "\n",
      "Could not run deepvariant as part of ont_germline pipeline\n",
      "Exiting pbrun ...\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all \\\n",
    "  --memory 60g \\\n",
    "  --ipc=host \\\n",
    "  -e CUDA_VISIBLE_DEVICES=\"0,1,2,3\" \\\n",
    "  -v $(pwd):/workdir \\\n",
    "  -w /workdir \\\n",
    "  nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 \\\n",
    "  pbrun ont_germline \\\n",
    "    --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "    --in-bam isolate_01_sorted.bam \\\n",
    "    --out-variants isolate_01_variants.vcf \\\n",
    "    --out-bam isolate_01_processed.bam \\\n",
    "    --low-memory \\\n",
    "    --run-partition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0e09919c-0dca-48df-a0d5-620255123aaa",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid decimal literal (3026747042.py, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[1], line 3\u001b[0;36m\u001b[0m\n\u001b[0;31m    --memory 60g \\\u001b[0m\n\u001b[0m              ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid decimal literal\n"
     ]
    }
   ],
   "source": [
    "docker run --rm \\\n",
    "  --gpus '\"device=0,1,2,3\"' \\\n",
    "  --memory 60g \\\n",
    "  --ipc=host \\\n",
    "  -v \"$(pwd)\":/workdir \\\n",
    "  -w /workdir \\\n",
    "  nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 \\\n",
    "  pbrun ont_germline \\\n",
    "    --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "    --in-bam isolate_01_sorted.bam \\\n",
    "    --out-variants isolate_01_variants.vcf \\\n",
    "    --out-bam isolate_01_processed.bam \\\n",
    "    --num-threads 32 \\\n",
    "    --num-streams-per-gpu 2 \\\n",
    "    --num-cpu-threads-per-stream 4 \\\n",
    "    --gpusort --gpuwrite \\\n",
    "    --low-memory \\\n",
    "    --run-partition \\\n",
    "    --gpu-num-per-partition 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "52d51b9c-147c-4253-acbd-2ba19bb55f22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[Parabricks Options Mesg]: Setting --num-streams-per-gpu based on available device memory.\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[PB Info 2025-Nov-19 14:32:15] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:32:15] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 14:32:15] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 14:32:15] ||                                 minimap2                                 ||\n",
      "[PB Info 2025-Nov-19 14:32:15] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:32:17] Reading reference file.\n",
      "[PB Info 2025-Nov-19 14:32:17] -------------------------------------\n",
      "[PB Info 2025-Nov-19 14:32:17] Elapsed-Minutes\tProcessed-Reads\n",
      "[PB Info 2025-Nov-19 14:32:17] -------------------------------------\n",
      "[PB Info 2025-Nov-19 14:32:17] \t0.0\t\t0\n",
      "[PB Info 2025-Nov-19 14:32:23] \t0.1\t\t0\n",
      "[PB Info 2025-Nov-19 14:32:29] \t0.2\t\t0\n",
      "[PB Info 2025-Nov-19 14:32:35] \t0.3\t\t0\n",
      "[PB Info 2025-Nov-19 14:32:41] \t0.4\t\t0\n",
      "[PB Info 2025-Nov-19 14:32:47] \t0.5\t\t0\n",
      "[PB Info 2025-Nov-19 14:32:53] \t0.6\t\t0\n",
      "[PB Info 2025-Nov-19 14:32:59] \t0.7\t\t5000\n",
      "[PB Info 2025-Nov-19 14:33:05] \t0.8\t\t5000\n",
      "[PB Info 2025-Nov-19 14:33:11] \t0.9\t\t5000\n",
      "[PB Info 2025-Nov-19 14:33:17] \t1.0\t\t20000\n",
      "[PB Info 2025-Nov-19 14:33:23] \t1.1\t\t20000\n",
      "[PB Info 2025-Nov-19 14:33:29] \t1.2\t\t25000\n",
      "[PB Info 2025-Nov-19 14:33:35] \t1.3\t\t30000\n",
      "[PB Info 2025-Nov-19 14:33:41] \t1.4\t\t30000\n",
      "[PB Info 2025-Nov-19 14:33:47] \t1.5\t\t35000\n",
      "[PB Info 2025-Nov-19 14:33:53] \t1.6\t\t45000\n",
      "[PB Info 2025-Nov-19 14:33:59] \t1.7\t\t55000\n",
      "[PB Info 2025-Nov-19 14:34:05] \t1.8\t\t65000\n",
      "[PB Info 2025-Nov-19 14:34:11] \t1.9\t\t85000\n",
      "[PB Info 2025-Nov-19 14:34:17] \t2.0\t\t100000\n",
      "[PB Info 2025-Nov-19 14:34:23] \t2.1\t\t120000\n",
      "[PB Info 2025-Nov-19 14:34:29] \t2.2\t\t125000\n",
      "[PB Info 2025-Nov-19 14:34:35] \t2.3\t\t135000\n",
      "[PB Info 2025-Nov-19 14:34:41] \t2.4\t\t150000\n",
      "[PB Info 2025-Nov-19 14:34:47] minimap2 finished in 152.077 seconds\n",
      "[PB Info 2025-Nov-19 14:34:47] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:34:47] ||        Program:                                          minimap2        ||\n",
      "[PB Info 2025-Nov-19 14:34:47] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 14:34:47] ||        Start Time:                       Wed Nov 19 14:32:15 2025        ||\n",
      "[PB Info 2025-Nov-19 14:34:47] ||        End Time:                         Wed Nov 19 14:34:47 2025        ||\n",
      "[PB Info 2025-Nov-19 14:34:47] ||        Total Time:                           2 minutes 32 seconds        ||\n",
      "[PB Info 2025-Nov-19 14:34:47] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:34:49] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:34:49] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 14:34:49] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 14:34:49] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2025-Nov-19 14:34:49] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:34:49] progressMeter - Percentage\n",
      "[PB Info 2025-Nov-19 14:34:49] 0.0\n",
      "[PB Info 2025-Nov-19 14:34:54] Sorting and Marking: 5.000 seconds\n",
      "[PB Info 2025-Nov-19 14:34:54] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:34:54] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2025-Nov-19 14:34:54] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 14:34:54] ||        Start Time:                       Wed Nov 19 14:34:49 2025        ||\n",
      "[PB Info 2025-Nov-19 14:34:54] ||        End Time:                         Wed Nov 19 14:34:54 2025        ||\n",
      "[PB Info 2025-Nov-19 14:34:54] ||        Total Time:                                      5 seconds        ||\n",
      "[PB Info 2025-Nov-19 14:34:54] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:34:54] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:34:54] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 14:34:54] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 14:34:54] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2025-Nov-19 14:34:54] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:34:54] Using PBBinBamFile for BAM writing\n",
      "[PB Info 2025-Nov-19 14:34:54] progressMeter -\tPercentage\n",
      "[PB Info 2025-Nov-19 14:35:04] 77.2\n",
      "[PB Info 2025-Nov-19 14:35:14] 100.0\n",
      "[PB Info 2025-Nov-19 14:35:18] BQSR and writing final BAM:  23.045 seconds\n",
      "[PB Info 2025-Nov-19 14:35:18] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:35:18] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2025-Nov-19 14:35:18] ||        Version:                                           4.6.0-1        ||\n",
      "[PB Info 2025-Nov-19 14:35:18] ||        Start Time:                       Wed Nov 19 14:34:54 2025        ||\n",
      "[PB Info 2025-Nov-19 14:35:18] ||        End Time:                         Wed Nov 19 14:35:18 2025        ||\n",
      "[PB Info 2025-Nov-19 14:35:18] ||        Total Time:                                     24 seconds        ||\n",
      "[PB Info 2025-Nov-19 14:35:18] ------------------------------------------------------------------------------\n",
      "Detected 4 CUDA Capable device(s), considering 4 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.9 / 12.9\n",
      "Using model for CUDA Capability Major/Minor version number:    75\n",
      "[PB Info 2025-Nov-19 14:35:22] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:35:22] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2025-Nov-19 14:35:22] ||                              Version 4.6.0-1                             ||\n",
      "[PB Info 2025-Nov-19 14:35:22] ||                                deepvariant                               ||\n",
      "[PB Info 2025-Nov-19 14:35:22] ------------------------------------------------------------------------------\n",
      "[PB Info 2025-Nov-19 14:35:22] Starting DeepVariant\n",
      "[PB Info 2025-Nov-19 14:35:22] Running with 2 GPU devices, each with 2 group instances and 6 workers\n",
      "[PB Info 2025-Nov-19 14:35:22] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2025-Nov-19 14:35:28] ProgressMeter -\tNC_000913.3:0\t0.1\n",
      "[PB Info 2025-Nov-19 14:35:34] ProgressMeter -\tNC_000913.3:0\t0.2\n",
      "[PB Info 2025-Nov-19 14:35:40] ProgressMeter -\tNC_000913.3:0\t0.3\n",
      "[PB Info 2025-Nov-19 14:35:46] ProgressMeter -\tNC_000913.3:0\t0.4\n",
      "[PB Info 2025-Nov-19 14:35:52] ProgressMeter -\tNC_000913.3:0\t0.5\n",
      "[PB Info 2025-Nov-19 14:35:58] ProgressMeter -\tNC_000913.3:0\t0.6\n",
      "[PB Info 2025-Nov-19 14:36:06] ProgressMeter -\tNC_000913.3:0\t0.7\n",
      "Killed\n",
      "CUDA_VISIBLE_DEVICES=0,1 /usr/local/parabricks/binaries/bin/deepsomatic 2 2 --ref /workdir/Ecoli_K12_MG1655_ref.fasta --reads /workdir/isolate_01_processed.bam -o 0.vcf  -L \"NC_000913.3\" -n 6 --model /usr/local/parabricks/binaries/model/75/ont/deepvariant.eng -long_reads --sort_by_haplotypes --add_hp_channel --parse_sam_aux_fields -norealign_reads --vsc_min_fraction_indels 0.12 --track_ref_reads --phase_reads --pileup_image_width 99 --max_reads_per_partition 600 --partition_size 25000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.08 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup diff_channels --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10 \n",
      "Process terminated with signal [SIGKILL: 9]. SIGKILL cannot be caught. A common reason for SIGKILL is running out of\n",
      "host memory. If the user has root access, they may be able to check by running `sudo journalctl -k --since \"<#> minutes\n",
      "ago\" | grep \"Killed process\"` to see the reason why processes were recently killed.\n",
      "Process exited with failure. Please check the logs.\n",
      "For technical support visit https://docs.nvidia.com/clara/index.html#parabricks\n",
      "Exiting...\n",
      "\n",
      "\n",
      "Could not run deepvariant as part of ont_germline pipeline\n",
      "Exiting pbrun ...\n"
     ]
    }
   ],
   "source": [
    "!docker run --rm --gpus all \\\n",
    "  --memory 60g \\\n",
    "  --ipc=host \\\n",
    "  -e CUDA_VISIBLE_DEVICES=\"0,1,2,3\" \\\n",
    "  -v $(pwd):/workdir \\\n",
    "  -w /workdir \\\n",
    "  nvcr.io/nvidia/clara/clara-parabricks:4.6.0-1 \\\n",
    "  pbrun ont_germline \\\n",
    "    --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "    --in-bam isolate_01_sorted.bam \\\n",
    "    --out-variants isolate_01_variants.vcf \\\n",
    "    --out-bam isolate_01_processed.bam \\\n",
    "    --low-memory \\\n",
    "    --run-partition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6a751ea3-9c58-4c0c-ac08-3b6362c5de50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: |=grep -iE \"image|container\"\n"
     ]
    }
   ],
   "source": [
    "env | grep -iE \"image|container\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f4b5e01d-127f-4ff5-a8b4-07ecf1fc0730",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: |=grep -iE \"image|container\"\n"
     ]
    }
   ],
   "source": [
    "env | grep -iE \"image|container\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "31c1315d-6303-467a-b785-a9c8b2adf5b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "pbrun: 4.6.0-1\n"
     ]
    }
   ],
   "source": [
    "!pbrun --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c315ea95-0707-4c47-8438-fd1cb44daafc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "[Parabricks Options Error]: The command argument ont-germlime is not a valid Parabricks tools/pipeline\n",
      "[Parabricks Options Error]: Run with -h to see help\n"
     ]
    }
   ],
   "source": [
    "!pbrun ont-germlime -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c64bf814-cadc-48d3-bef8-65b51c501c37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools 1.19.2\n",
      "Using htslib 1.19.1\n",
      "Copyright (C) 2024 Genome Research Ltd.\n",
      "\n",
      "Samtools compilation details:\n",
      "    Features:       build=configure curses=yes \n",
      "    CC:             gcc\n",
      "    CPPFLAGS:       \n",
      "    CFLAGS:         -Wall -g -O2\n",
      "    LDFLAGS:        \n",
      "    HTSDIR:         htslib-1.19.1\n",
      "    LIBS:           \n",
      "    CURSES_LIB:     -lncursesw\n",
      "\n",
      "HTSlib compilation details:\n",
      "    Features:       build=configure libcurl=yes S3=yes GCS=yes libdeflate=no lzma=yes bzip2=yes plugins=no htscodecs=1.6.0\n",
      "    CC:             gcc\n",
      "    CPPFLAGS:       \n",
      "    CFLAGS:         -Wall -g -O2 -fvisibility=hidden\n",
      "    LDFLAGS:        -fvisibility=hidden \n",
      "\n",
      "HTSlib URL scheme handlers present:\n",
      "    built-in:\t preload, data, file\n",
      "    S3 Multipart Upload:\t s3w, s3w+https, s3w+http\n",
      "    Amazon S3:\t s3, s3+https, s3+http\n",
      "    Google Cloud Storage:\t gs+http, gs, gs+https\n",
      "    libcurl:\t ftp, http, https, ftps\n",
      "    crypt4gh-needed:\t crypt4gh\n",
      "    mem:\t mem\n"
     ]
    }
   ],
   "source": [
    "!samtools --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b2f73187-41e8-427a-9469-b9b6c34ec3bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using GATK jar /usr/local/gatk/gatk-package-4.6.1.0-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /usr/local/gatk/gatk-package-4.6.1.0-local.jar --version\n",
      "The Genome Analysis Toolkit (GATK) v4.6.1.0\n",
      "HTSJDK Version: 4.1.3\n",
      "Picard Version: 3.3.0\n"
     ]
    }
   ],
   "source": [
    "!gatk --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9112e138-29ad-46c4-9c2e-1aca78ba1b28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bcftools 1.19\n",
      "Using htslib 1.19\n",
      "Copyright (C) 2023 Genome Research Ltd.\n",
      "License Expat: The MIT/Expat license\n",
      "This is free software: you are free to change and redistribute it.\n",
      "There is NO WARRANTY, to the extent permitted by law.\n"
     ]
    }
   ],
   "source": [
    "!bcftools --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1ee8d68f-f109-44c7-b75b-cffebb56372e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Nov 19 21:51:14 2025       \n",
      "+-----------------------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |\n",
      "+-----------------------------------------+------------------------+----------------------+\n",
      "| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |\n",
      "|                                         |                        |               MIG M. |\n",
      "|=========================================+========================+======================|\n",
      "|   0  NVIDIA A100-SXM4-40GB          On  |   00000000:A0:1C.0 Off |                    0 |\n",
      "| N/A   50C    P0             62W /  400W |       0MiB /  40960MiB |      0%      Default |\n",
      "|                                         |                        |             Disabled |\n",
      "+-----------------------------------------+------------------------+----------------------+\n",
      "|   1  NVIDIA A100-SXM4-40GB          On  |   00000000:A0:1D.0 Off |                    0 |\n",
      "| N/A   45C    P0             59W /  400W |       0MiB /  40960MiB |      0%      Default |\n",
      "|                                         |                        |             Disabled |\n",
      "+-----------------------------------------+------------------------+----------------------+\n",
      "\n",
      "+-----------------------------------------------------------------------------------------+\n",
      "| Processes:                                                                              |\n",
      "|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |\n",
      "|        ID   ID                                                               Usage      |\n",
      "|=========================================================================================|\n",
      "|  No running processes found                                                             |\n",
      "+-----------------------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "854fa9a8-fc39-4f81-9049-5f35804d35dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "usage: pbrun ont_germline <options>\n",
      "Help: pbrun ont_germline -h\n",
      "\n",
      "Run the germline pipeline from FASTQ/BAM to VCF by aligning long read ONT\n",
      "sequences with minimap2 and using a deep neural network analysis.\n",
      "\n",
      "options:\n",
      "  -h, --help             show this help message and exit\n",
      "\n",
      "Input Output file options:\n",
      "  Options for Input and Output files for this tool.\n",
      "\n",
      "  --ref REF              Path to the reference file.\n",
      "  --index INDEX          Path to a minimizer index file generated by vanilla minimap2\n",
      "                         to reduce indexing time.\n",
      "  --in-fq IN_FQ          Path to a query sequence file in fastq or fastq.gz format.\n",
      "  --in-bam IN_BAM        Path to the input BAM/CRAM file.\n",
      "  --knownSites KNOWNSITES\n",
      "                         Path to a known indels file. The file must be in vcf.gz\n",
      "                         format. This option can be used multiple times.\n",
      "  --interval-file INTERVAL_FILE\n",
      "                         Path to an interval file in one of these formats: Picard-style\n",
      "                         (.interval_list or .picard), GATK-style (.list or .intervals),\n",
      "                         or BED file (.bed). This option can be used multiple times.\n",
      "  --pb-model-file PB_MODEL_FILE\n",
      "                         Path to a non-default parabricks model file for deepvariant.\n",
      "  --out-recal-file OUT_RECAL_FILE\n",
      "                         Path of the report file after Base Quality Score Recalibration.\n",
      "  --out-bam OUT_BAM      Path of BAM file after marking duplicates.\n",
      "  --out-variants OUT_VARIANTS\n",
      "                         Path of the vcf/vcf.gz/gvcf/gvcf.gz file after variant\n",
      "                         calling.\n",
      "  --out-duplicate-metrics OUT_DUPLICATE_METRICS\n",
      "                         Path of a duplicate metrics file after marking duplicates.\n",
      "  --proposed-variants PROPOSED_VARIANTS\n",
      "                         Path of the VCF file, which has proposed variants for the\n",
      "                         make examples stage.\n",
      "\n",
      "Tool Options:\n",
      "  Options specific to the tool.\n",
      "\n",
      "  -k MINIMIZER_KMER_LEN, --minimizer-kmer-len MINIMIZER_KMER_LEN\n",
      "                         Minimizer k-mer length.\n",
      "  -uf, --forward-transcript-strand\n",
      "                         Force minimap2 to consider the forward transcript strand\n",
      "                         only when finding canonical splicing sites GT-AG.\n",
      "  --eqx                  Write =/X CIGAR operators.\n",
      "  -L INTERVAL, --interval INTERVAL\n",
      "                         Interval within which to call bqsr from the input reads.\n",
      "                         All intervals will have a padding of 100 to get read records,\n",
      "                         and overlapping intervals will be combined. Interval files\n",
      "                         should be passed using the --interval-file option. This\n",
      "                         option can be used multiple times (e.g. \"-L chr1 -L chr2:10000\n",
      "                         -L chr3:20000+ -L chr4:10000-20000\").\n",
      "  -ip INTERVAL_PADDING, --interval-padding INTERVAL_PADDING\n",
      "                         Amount of padding (in base pairs) to add to each interval\n",
      "                         you are including.\n",
      "  --standalone-bqsr      Run standalone BQSR after generating sorted BAM. This option\n",
      "                         requires both --knownSites and --out-recal-file input parameters.\n",
      "  --read-group-sm READ_GROUP_SM\n",
      "                         SM tag for read groups in this run.\n",
      "  --read-group-lb READ_GROUP_LB\n",
      "                         LB tag for read groups in this run.\n",
      "  --read-group-pl READ_GROUP_PL\n",
      "                         PL tag for read groups in this run.\n",
      "  --read-group-id-prefix READ_GROUP_ID_PREFIX\n",
      "                         Prefix for the ID and PU tags for read groups in this run.\n",
      "                         This prefix will be used for all pairs of FASTQ files in\n",
      "                         this run. The ID and PU tags will consist of this prefix\n",
      "                         and an identifier, that will be unique for a pair of FASTQ\n",
      "                         files.\n",
      "  --disable-use-window-selector-model\n",
      "                         Change the window selector model from Allele Count Linear\n",
      "                         to Variant Reads. This option will increase the accuracy\n",
      "                         and runtime.\n",
      "  --gvcf                 Generate variant calls in .gvcf format.\n",
      "  --norealign-reads      Do not locally realign reads before calling variants. Reads\n",
      "                         longer than 500 bp are never realigned.\n",
      "  --sort-by-haplotypes   Reads are sorted by haplotypes (using HP tag).\n",
      "  --keep-duplicates      Keep reads that are duplicate.\n",
      "  --keep-legacy-allele-counter-behavior\n",
      "                         If specified, the behavior in this commit is reverted:\n",
      "                         'https://github.com/google/deepvariant/commit/fbde0674639a28cb9e8004c7a01bbe25240c7d46'.\n",
      "                         We do not recommend setting this flag to True.\n",
      "  --vsc-min-count-snps VSC_MIN_COUNT_SNPS\n",
      "                         SNP alleles occurring at least this many times in the AlleleCount\n",
      "                         will be advanced as candidates. (default: 2)\n",
      "  --vsc-min-count-indels VSC_MIN_COUNT_INDELS\n",
      "                         Indel alleles occurring at least this many times in the\n",
      "                         AlleleCount will be advanced as candidates. (default: 2)\n",
      "  --vsc-min-fraction-snps VSC_MIN_FRACTION_SNPS\n",
      "                         SNP alleles occurring at least this fraction of all counts\n",
      "                         in the AlleleCount will be advanced as candidates. (default:\n",
      "                         0.12)\n",
      "  --vsc-min-fraction-indels VSC_MIN_FRACTION_INDELS\n",
      "                         Indel alleles occurring at least this fraction of all counts\n",
      "                         in the AlleleCount will be advanced as candidates.\n",
      "  --min-mapping-quality MIN_MAPPING_QUALITY\n",
      "                         By default, reads with any mapping quality are kept. Setting\n",
      "                         this field to a positive integer i will only keep reads\n",
      "                         that have a MAPQ >= i. Note this only applies to aligned\n",
      "                         reads. (default: 5)\n",
      "  --min-base-quality MIN_BASE_QUALITY\n",
      "                         Minimum base quality. This option enforces a minimum base\n",
      "                         quality score for alternate alleles. Alternate alleles\n",
      "                         will only be considered if all bases in the allele have\n",
      "                         a quality greater than min_base_quality. (default: 10)\n",
      "  --alt-aligned-pileup ALT_ALIGNED_PILEUP\n",
      "                         Value can be one of [none, diff_channels]. Include alignments\n",
      "                         of reads against each candidate alternate allele in the\n",
      "                         pileup image.\n",
      "  --variant-caller VARIANT_CALLER\n",
      "                         Value can be one of [VERY_SENSITIVE_CALLER, VCF_CANDIDATE_IMPORTER].\n",
      "                         The caller to use to make examples. If you use VCF_CANDIDATE_IMPORTER,\n",
      "                         it implies force calling. Default is VERY_SENSITIVE_CALLER.\n",
      "  --add-hp-channel       Add another channel to represent HP tags per read.\n",
      "  --parse-sam-aux-fields\n",
      "                         Auxiliary fields of the BAM/CRAM records are parsed. If\n",
      "                         either --sort-by-haplotypes or --add-hp-channel is set,\n",
      "                         then this option must also be set.\n",
      "  --use-wes-model        If specified, the WES model file will be used. Only used\n",
      "                         in shortread mode.\n",
      "  --include-med-dp       If specified, include MED_DP in the output gVCF records.\n",
      "  --normalize-reads      If specified, allele counter left align INDELs for each\n",
      "                         read.\n",
      "  --pileup-image-width PILEUP_IMAGE_WIDTH\n",
      "                         Pileup image width. Only change this if you know your model\n",
      "                         supports this width. (default: 221)\n",
      "  --channel-insert-size\n",
      "                         If specified, add insert_size channel into the pileup image.\n",
      "                         By default, this parameter is true in WGS and WES mode.\n",
      "  --no-channel-insert-size\n",
      "                         If specified, don't add insert_size channel into the pileup\n",
      "                         image.\n",
      "  --max-read-size-512    Allow deepvariant to run on reads of size 512bp. The default\n",
      "                         size is 320 bp.\n",
      "  --prealign-helper-thread\n",
      "                         Use an extra thread for the pre-align step. This parameter\n",
      "                         is more useful when --max-reads-size-512 is set.\n",
      "  --track-ref-reads      If specified, allele counter keeps track of reads supporting\n",
      "                         ref. By default, allele counter keeps a simple count of\n",
      "                         the number of reads supporting ref.\n",
      "  --phase-reads          Calculate phases and add HP tag to all reads automatically.\n",
      "  --dbg-min-base-quality DBG_MIN_BASE_QUALITY\n",
      "                         Minimum base quality in a k-mer sequence to consider. (default:\n",
      "                         15)\n",
      "  --ws-min-windows-distance WS_MIN_WINDOWS_DISTANCE\n",
      "                         Minimum distance between candidate windows for local assembly.\n",
      "                         (default: 80)\n",
      "  --channel-gc-content   If specified, add gc_content channel into the pileup image.\n",
      "  --channel-hmer-deletion-quality\n",
      "                         If specified, add hmer deletion quality channel into the\n",
      "                         pileup image.\n",
      "  --channel-hmer-insertion-quality\n",
      "                         If specified, add hmer insertion quality channel into the\n",
      "                         pileup image.\n",
      "  --channel-non-hmer-insertion-quality\n",
      "                         If specified, add non-hmer insertion quality channel into\n",
      "                         the pileup image.\n",
      "  --skip-bq-channel      If specified, ignore base quality channel.\n",
      "  --aux-fields-to-keep AUX_FIELDS_TO_KEEP\n",
      "                         Comma-delimited list of auxiliary BAM fields to keep. Values\n",
      "                         can be [HP, tp, t0]. (default: HP)\n",
      "  --vsc-min-fraction-hmer-indels VSC_MIN_FRACTION_HMER_INDELS\n",
      "                         Hmer Indel alleles occurring at least this be advanced\n",
      "                         as candidates. Use this threshold if hmer and non-hmer\n",
      "                         indels should be treated differently (Ultima reads)Default\n",
      "                         will use the same threshold for hmer and non-hmer indels,\n",
      "                         as defined in vsc_min_fraction_indels.\n",
      "  --vsc-turn-on-non-hmer-ins-proxy-support\n",
      "                         Add read-support from soft-clipped reads and other non-hmer\n",
      "                         insertion alleles,to the most frequent non-hmer insertion\n",
      "                         allele.\n",
      "  --consider-strand-bias\n",
      "                         If specified, expect SB field in calls and write it to\n",
      "                         the VCF file.\n",
      "  --p-error P_ERROR      Basecalling error for reference confidence model. (default:\n",
      "                         0.001)\n",
      "  --channel-ins-size     If specified, add another channel to represent size of\n",
      "                         insertions (good for flow-based sequencing).\n",
      "  --max-ins-size MAX_INS_SIZE\n",
      "                         Max insertion size for ins_size_channel, larger insertions\n",
      "                         will look like max (have max intensity). (default: 10)\n",
      "  --disable-group-variants\n",
      "                         If using vcf_candidate_importer and multi-allelic sites\n",
      "                         are split across multiple lines in VCF, add this flag so\n",
      "                         that variants are not grouped when transforming CallVariantsOutput\n",
      "                         to Variants.\n",
      "  --filter-reads-too-long\n",
      "                         Ignore all input BAM reads with size > 512bp.\n",
      "  --haploid-contigs HAPLOID_CONTIGS\n",
      "                         Optional list of non autosomal chromosomes. For all listed\n",
      "                         chromosomes HET probabilities are not considered.\n",
      "\n",
      "Performance Options:\n",
      "  Performance Tuning Options.\n",
      "\n",
      "  --num-threads NUM_THREADS\n",
      "                         Number of processing threads. (default: 88)\n",
      "  --nstreams NSTREAMS    Number of streams to use per GPU. (default: 2)\n",
      "  --gpuwrite             Use one GPU to accelerate writing final BAM/CRAM.\n",
      "  --gpuwrite-deflate-algo GPUWRITE_DEFLATE_ALGO\n",
      "                         Choose the nvCOMP DEFLATE algorithm to use with --gpuwrite.\n",
      "                         Note these options do not correspond to CPU DEFLATE options.\n",
      "                         Valid options are 1, 2, and 4. Option 1 is fastest, while\n",
      "                         options 2 and 4 have progressively lower throughput but\n",
      "                         higher compression ratios. The default value is 1 when\n",
      "                         the user does not provide an input (i.e., None).\n",
      "  --gpusort              Use GPUs to accelerate sorting.\n",
      "  --use-gds              Use GPUDirect Storage (GDS) to enable a direct data path\n",
      "                         for direct memory access (DMA) transfers between GPU memory\n",
      "                         and storage. Must be used concurrently with `--gpuwrite`.\n",
      "                         Please refer to Parabricks Documentation > Best Performance\n",
      "                         for information on how to set up and use GPUDirect Storage.\n",
      "  --max-queue-reads MAX_QUEUE_READS\n",
      "                         Max number of reads to allow in the alignment processing\n",
      "                         stage. Increasing this value may result in faster processing,\n",
      "                         but it will use more host memory. (default: 500000)\n",
      "  --low-memory           Use low memory mode.\n",
      "  --chunk-size CHUNK_SIZE\n",
      "                         Max number of reads in a processing chunk. Increasing this\n",
      "                         value may result in faster processing for splice presets,\n",
      "                         but it will use more host memory. (default: 5000)\n",
      "  --num-cpu-threads-per-stream NUM_CPU_THREADS_PER_STREAM\n",
      "                         Number of CPU threads to use per stream. (default: 6)\n",
      "  --num-streams-per-gpu NUM_STREAMS_PER_GPU\n",
      "                         Number of streams to use per GPU. Default is 'auto' which\n",
      "                         will try to use an optimal amount of streams based on the\n",
      "                         GPU. (default: auto)\n",
      "  --run-partition        Divide the whole genome into multiple partitions and run\n",
      "                         multiple processes at the same time, each on one partition.\n",
      "  --gpu-num-per-partition GPU_NUM_PER_PARTITION\n",
      "                         Number of GPUs to use per partition.\n",
      "  --max-reads-per-partition MAX_READS_PER_PARTITION\n",
      "                         The maximum number of reads per partition that are considered\n",
      "                         before following processing such as sampling and realignment.\n",
      "                         (default: 1500)\n",
      "  --partition-size PARTITION_SIZE\n",
      "                         The maximum number of basepairs allowed in a region before\n",
      "                         splitting it into multiple smaller subregions. (default:\n",
      "                         1000)\n",
      "  --use-tf32             Enable inference optimization using Tensor Float 32(TF32)\n",
      "                         on ampere+ gpu. Note that this might introduce a few mismatches\n",
      "                         in the output VCF.\n",
      "  --read-from-tmp-dir    Running variant caller reading from bin files generated\n",
      "                         by Aligner and sort. Run postsort in parallel. This option\n",
      "                         will increase device memory usage.\n",
      "\n",
      "Run Options:\n",
      "  Options related to resource/verbosity/etc.\n",
      "\n",
      "  --verbose              Enable verbose output.\n",
      "  --x3                   Show full command line arguments.\n",
      "  --logfile LOGFILE      Path to the log file. If not specified, messages will only\n",
      "                         be written to the standard error output.\n",
      "  --tmp-dir TMP_DIR      Full path to the directory where temporary files will be\n",
      "                         stored. (default: .)\n",
      "  --with-petagene-dir WITH_PETAGENE_DIR\n",
      "                         Full path to the PetaGene installation directory. By default,\n",
      "                         this should have been installed at /opt/petagene. Use of\n",
      "                         this option also requires that the PetaLink library has\n",
      "                         been preloaded by setting the LD_PRELOAD environment variable.\n",
      "                         Optionally set the PETASUITE_REFPATH and PGCLOUD_CREDPATH\n",
      "                         environment variables that are used for data and credentials.\n",
      "                         Optionally set the PetaLinkMode environment variable that\n",
      "                         is used to further configure PetaLink, notably setting\n",
      "                         it to \"+write\" to enable outputting compressed BAM and\n",
      "                         .fastq files.\n",
      "  --keep-tmp             Do not delete the directory storing temporary files after\n",
      "                         completion.\n",
      "  --no-seccomp-override\n",
      "                         Do not override seccomp options for docker.\n",
      "  --version              View compatible software versions.\n",
      "  --preserve-file-symlinks\n",
      "                         Override default behavior to keep file symlinks intact\n",
      "                         and *not* resolve the symlink.\n",
      "  --num-gpus NUM_GPUS    Number of GPUs to use for a run. (default: 2)\n"
     ]
    }
   ],
   "source": [
    "!pbrun ont_germline -h\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a600255e-904b-4b3f-8e10-a52046242c8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gunzip Parabricks_Pathogentest_Parabricks_pathogen_Test_isolate_01_sorted.bam.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "490fcf01-5a5c-4db2-9c0d-181b99449657",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /opt/app-root/src/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "[Parabricks Options Mesg]: Setting --num-streams-per-gpu based on available device memory.\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Read group created for /opt/app-root/src/isolate_01_sorted.bam\n",
      "[Parabricks Options Mesg]: @RG\\tID:SRR15462566.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:SRR15462566.1\n",
      "/opt/parabricks/binaries/bin/minimap2: error while loading shared libraries: libtcmalloc_minimal.so.4: cannot open shared object file: No such file or directory\n",
      "Process exited with error code: 127.\n",
      "For technical support visit https://docs.nvidia.com/clara/index.html#parabricks\n",
      "Exiting...\n",
      "\n",
      "Could not run minimap2 as part of ont_germline pipeline\n",
      "Exiting pbrun ...\n"
     ]
    }
   ],
   "source": [
    "!CUDA_VISIBLE_DEVICES=\"0,1\" \\\n",
    "pbrun ont_germline \\\n",
    "  --ref Ecoli_K12_MG1655_ref.fasta \\\n",
    "  --in-bam isolate_01_sorted.bam \\\n",
    "  --out-variants isolate_01_variants.vcf \\\n",
    "  --out-bam isolate_01_processed.bam \\\n",
    "  --low-memory \\\n",
    "  --run-partition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5174585b-b14c-4c43-be74-3a2fb01df18b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "conda-base-py",
   "name": "workbench-notebooks.m130",
   "type": "gcloud",
   "uri": "us-docker.pkg.dev/deeplearning-platform-release/gcr.io/workbench-notebooks:m130"
  },
  "kernelspec": {
   "display_name": "Python 3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
